<?php
/*******************************************************************************
 *
 * LEIDEN OPEN VARIATION DATABASE (LOVD)
 *
 * Created     : 2009-10-22
 * Modified    : 2016-04-05
 * For LOVD    : 3.0-14
 *
 * Copyright   : 2004-2016 Leiden University Medical Center; http://www.LUMC.nl/
 * Programmers : Ing. Ivo F.A.C. Fokkema <I.F.A.C.Fokkema@LUMC.nl>
 *               Ing. Ivar C. Lugtenburg <I.C.Lugtenburg@LUMC.nl>
 *
 *
 * This file is part of LOVD.
 *
 * LOVD is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * LOVD is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with LOVD.  If not, see <http://www.gnu.org/licenses/>.
 *
 *************/

// DMD_SPECIFIC
if (!defined('ROOT_PATH')) {
    define('ROOT_PATH', '../');
    require ROOT_PATH . 'inc-init.php';
}

if (strtoupper($_INI['instance']['name']) == 'MGHA') {
    // Site specific columns for Melbourne Genomics Health Alliance (MGHA).
    $aColSQL = array (

        // INDIVIDUAL & SCREENING COLUMNS
        //                                      COL ID,                                   ORDER,WIDTH,HGVS,STND,MAND,      TITLE,                   FORM DESCRIPTION, SHORT LEGEND,  FULL LEGEND,  MYSQL,  FORM DEFINITION  SELECT OPTIONS, REGEXP,  PUBLIC (VIEW,ADD), COUNT?, CREATED_BY, CREATED DATE, EDITED BY, EDITED DATE.
        'INSERT INTO ' . TABLE_COLS . ' VALUES ("Phenotype/Age",                               10, 100, 0, 0, 0, "Age examined",         "Type 35y for 35 years, 04y08m for 4 years and 8 months, 18y? for around 18 years, >54y for older than 54, ? for unknown.", "The age at which the individual was examined, if known. 04y08m = 4 years and 8 months.", "The age at which the individual was examined, if known.\r\n<UL style=\"margin-top:0px;\">\r\n  <LI>35y = 35 years</LI>\r\n  <LI>04y08m = 4 years and 8 months</LI>\r\n  <LI>18y? = around 18 years</LI>\r\n  <LI>&gt;54y = older than 54</LI>\r\n  <LI>? = unknown</LI>\r\n</UL>", "VARCHAR(12)", "Age at examination|The age at which the individual was examined, if known. Numbers lower than 10 should be prefixed by a zero and the field should always begin with years, to facilitate sorting on this column.|text|10", "", "/^([<>]?\\\\d{2,3}y(\\\\d{2}m(\\\\d{2}d)?)?)?\\\\??$/", 1, 1, 1, 0, NOW(), NULL, NULL)',
        'INSERT INTO ' . TABLE_COLS . ' VALUES ("Individual/Reference",                         2, 200, 0, 0, 0, "Reference",            "", "Reference to publication describing the individual/family.", "Reference to publication describing the individual/family, possibly giving more phenotypic details than listed in this database entry, including link to PubMed or other source, e.g. \"den Dunnen ASHG2003 P2346\". References in the &quot;Country:City&quot; format indicate that the variant was submitted directly to this database by the laboratory indicated.", "VARCHAR(200)", "Reference||text|50", "", "", 1, 1, 1, 0, NOW(), NULL, NULL)',

        'INSERT INTO ' . TABLE_COLS . ' VALUES ("Screening/Pipeline/Run_ID",              255, 100, 0, 1, 0, "Pipeline Run ID",                     "", "Pipeline Run ID", "Pipeline Run ID", "VARCHAR(255)", "Pipeline Run ID||text|6", "", "", 1, 1, 1, 0, NOW(), NULL, NULL)',
        'INSERT INTO ' . TABLE_COLS . ' VALUES ("Individual/Sample_ID",                   1, 80, 0, 1, 0, "Sample ID",                              "", "The individual\'s ID in the hospital, diagnostic laboratory or in the publication.", "The individual\'s ID in the hospital, diagnostic laboratory or in the publication.", "VARCHAR(20)", "Sample ID||text|20", "", "", 1, 1, 1, 0, NOW(), NULL, NULL)',
        'INSERT INTO ' . TABLE_COLS . ' VALUES ("Individual/Gender",                      3,  70, 0, 1, 0, "Gender",                                "", "Individual\'s gender.", "The gender of the reported individual.", "VARCHAR(100)", "Gender||select|1|--Not specified--|false|false", "? = Unknown\r\nF = Female\r\nM = Male\r\nrF = Raised as female\r\nrM = Raised as male", "", 1, 1, 1, 0, NOW(), NULL, NULL)',
        'INSERT INTO ' . TABLE_COLS . ' VALUES ("Individual/Origin/Ethnic",               201, 200, 0, 1, 0, "Ethnic Origin",                       "Ethnic origin of individual; e.g. African, Caucasian, gypsy, jew (Ashkenazi).", "Ethnic origin of individual; e.g. African, Caucasian, gypsy, jew (Ashkenazi).", "The ethnic origin of the individual; e.g. African, Caucasian, gypsy, jew (Ashkenazi).", "VARCHAR(50)", "Ethnic origin|If mixed, please indicate origin of father and mother, if known.|text|20", "", "", 1, 1, 1, 0, NOW(), NULL, NULL)',
        'INSERT INTO ' . TABLE_COLS . ' VALUES ("Individual/Consanguinity",               240,  40, 0, 1, 0, "Consanguinity",                       "Indicates whether the parents are related (consanguineous), not related (non-consanguineous) or whether consanguinity is not known (unknown)", "Indicates whether the parents are related (consanguineous), not related (non-consanguineous) or whether consanguinity is not known (unknown)", "Indicates whether the parents are related (consanguineous), not related (non-consanguineous) or whether consanguinity is not known (unknown)", "VARCHAR(100)", "Consanguinity||select|1|--Not specified--|false|false", "? = Unknown\r\nno = Non-consanguineous parents\r\nyes = Consanguineous parents", "", 1, 1, 1, 0, NOW(), NULL, NULL)',
        'INSERT INTO ' . TABLE_COLS . ' VALUES ("Individual/Cohort",                      240,  80, 0, 1, 0, "Cohort",                              "", "Cohort", "Cohort", "VARCHAR(20)", "Cohort||text|8", "", "", 1, 1, 1, 0, NOW(), NULL, NULL)',
        'INSERT INTO ' . TABLE_COLS . ' VALUES ("Individual/Hospital_centre",             240,  80, 0, 1, 0, "Hospital Centre",                     "", "Hospital Centre", "Hospital Centre", "VARCHAR(20)", "Hospital Centre||text|8", "", "", 1, 1, 1, 0, NOW(), NULL, NULL)',
        'INSERT INTO ' . TABLE_COLS . ' VALUES ("Screening/Mother/Sample_ID",             1, 80, 0, 1, 0, "Mother\'s Sample ID",                     "", "The mother\'s ID in the hospital, diagnostic laboratory or in the publication.", "The mother\'s ID in the hospital, diagnostic laboratory or in the publication.", "VARCHAR(20)", "Mother\'s Sample ID||text|20", "", "", 1, 1, 1, 0, NOW(), NULL, NULL)',
        'INSERT INTO ' . TABLE_COLS . ' VALUES ("Screening/Mother/Origin/Ethnic",         201, 200, 0, 1, 0, "Mother\'s Ethnic Origin",              "Ethnic origin of mother; e.g. African, Caucasian, gypsy, jew (Ashkenazi).", "Ethnic origin of mother; e.g. African, Caucasian, gypsy, jew (Ashkenazi).", "The ethnic origin of the mother; e.g. African, Caucasian, gypsy, jew (Ashkenazi).", "VARCHAR(50)", "Mother\'s Ethnic origin|If mixed, please indicate origin of father and mother, if known.|text|20", "", "", 1, 1, 1, 0, NOW(), NULL, NULL)',
        'INSERT INTO ' . TABLE_COLS . ' VALUES ("Screening/Father/Sample_ID",             1, 80, 0, 1, 0, "Father\' Sample ID",                      "", "The father\'s ID in the hospital, diagnostic laboratory or in the publication.", "The father\'s ID in the hospital, diagnostic laboratory or in the publication.", "VARCHAR(20)", "Father\'s Sample ID||text|20", "", "", 1, 1, 1, 0, NOW(), NULL, NULL)',
        'INSERT INTO ' . TABLE_COLS . ' VALUES ("Screening/Father/Origin/Ethnic",         201, 200, 0, 1, 0, "Father\'s Ethnic Origin",              "Ethnic origin of father; e.g. African, Caucasian, gypsy, jew (Ashkenazi).", "Ethnic origin of father; e.g. African, Caucasian, gypsy, jew (Ashkenazi).", "The ethnic origin of the father; e.g. African, Caucasian, gypsy, jew (Ashkenazi).", "VARCHAR(50)", "Father\'s Ethnic origin|If mixed, please indicate origin of father and mother, if known.|text|20", "", "", 1, 1, 1, 0, NOW(), NULL, NULL)',
        'INSERT INTO ' . TABLE_COLS . ' VALUES ("Screening/Pipeline/Notes",               240,  200, 0, 1, 0, "Pipeline Notes",                     "", "Pipeline Notes", "Pipeline Notes", "TEXT", "Pipeline Notes||textarea|50|3", "", "", 1, 1, 1, 0, NOW(), NULL, NULL)',
        'INSERT INTO ' . TABLE_COLS . ' VALUES ("Screening/Notes",                        240,  200, 0, 1, 0, "Notes",                              "", "Notes", "Screening Notes", "TEXT", "Notes||textarea|50|3", "", "", 1, 1, 1, 0, NOW(), NULL, NULL)',
        'INSERT INTO ' . TABLE_COLS . ' VALUES ("Screening/Mother/Notes",                 240,  200, 0, 1, 0, "Mother\'s Screening Notes",          "", "Mother\'s Screening Notes", "Mother\'s Screening Notes", "TEXT", "Mother\'s Notes||textarea|50|3", "", "", 1, 1, 1, 0, NOW(), NULL, NULL)',
        'INSERT INTO ' . TABLE_COLS . ' VALUES ("Screening/Father/Notes",                 240,  200, 0, 1, 0, "Father\'s Screening Notes",          "", "Father\'s Screening Notes", "Father\'s Screening Notes", "TEXT", "Father\'s Notes||textarea|50|3", "", "", 1, 1, 1, 0, NOW(), NULL, NULL)',
        'INSERT INTO ' . TABLE_COLS . ' VALUES ("Screening/Batch",                        255, 100, 0, 1, 0, "Batch",                               "", "Batch", "Batch", "VARCHAR(255)", "Batch||text|6", "", "", 1, 1, 1, 0, NOW(), NULL, NULL)',
        'INSERT INTO ' . TABLE_COLS . ' VALUES ("Screening/DNA/Tube_ID",                  5, 80, 0, 1, 0, "DNA Tube ID",                            "", "The ID of the Individual\s DNA tube in the hospital or diagnostic laboratory.", "The ID of the Individual\'s DNA tube in the hospital or diagnostic laboratory.", "VARCHAR(20)", "DNA Tube ID||text|20", "", "", 1, 1, 1, 0, NOW(), NULL, NULL)',
        'INSERT INTO ' . TABLE_COLS . ' VALUES ("Screening/Mother/DNA/Tube_ID",           5, 80, 0, 1, 0, "Mother\'s DNA Tube ID",                  "", "The ID of the mother\'s DNA tube in the hospital or diagnostic laboratory.", "The ID of the mother\'s DNA tube in the hospital or diagnostic laboratory.", "VARCHAR(20)", "Mother\'s DNA Tube ID||text|20", "", "", 1, 1, 1, 0, NOW(), NULL, NULL)',
        'INSERT INTO ' . TABLE_COLS . ' VALUES ("Screening/Father/DNA/Tube_ID",           5, 80, 0, 1, 0, "Father\'s DNA Tube ID",                  "", "The ID of the father\'s DNA tube in the hospital or diagnostic laboratory.", "The ID of the father\'s DNA tube in the hospital or diagnostic laboratory.", "VARCHAR(20)", "Father\'s DNA Tube ID||text|20", "", "", 1, 1, 1, 0, NOW(), NULL, NULL)',
        'INSERT INTO ' . TABLE_COLS . ' VALUES ("Screening/DNA/Concentration",            100, 80, 0, 1, 0, "DNA Concentration",                    "", "The concentration of the DNA.", "The concentration of the DNA.", "MEDIUMINT(8)", "DNA Concentration||text|8", "", "", 1, 1, 1, 0, NOW(), NULL, NULL)',
        'INSERT INTO ' . TABLE_COLS . ' VALUES ("Screening/DNA/Volume",                   100, 80, 0, 1, 0, "DNA Volume",                           "", "The volume of DNA.", "The volume of DNA.", "SMALLINT(3)", "DNA Volume||text|8", "", "", 1, 1, 1, 0, NOW(), NULL, NULL)',
        'INSERT INTO ' . TABLE_COLS . ' VALUES ("Screening/DNA/Quantity",                 100, 80, 0, 1, 0, "DNA Quantity",                         "", "The quantity of DNA.", "The quantity of DNA.", "MEDIUMINT(8)", "DNA Quantity||text|8", "", "", 1, 1, 1, 0, NOW(), NULL, NULL)',
        'INSERT INTO ' . TABLE_COLS . ' VALUES ("Screening/DNA/Quality",                  100, 80, 0, 1, 0, "DNA Quality",                          "", "The quality of the DNA.", "The quantity of the DNA.", "VARCHAR(10)", "DNA Quality||text|8", "", "", 1, 1, 1, 0, NOW(), NULL, NULL)',
        'INSERT INTO ' . TABLE_COLS . ' VALUES ("Screening/DNA/Date",                     100, 80, 0, 1, 0, "DNA Date",                             "", "The date the DNA was extracted.", "The date the DNA was extracted.", "VARCHAR(255)", "DNA Date||text|10", "", "", 1, 1, 1, 0, NOW(), NULL, NULL)',
        'INSERT INTO ' . TABLE_COLS . ' VALUES ("Screening/Sample/Type",                  100, 80, 0, 1, 0, "Sample Type",                          "", "Sample Type", "Sample Type", "VARCHAR(20)", "Sample Type||text|8", "", "", 1, 1, 1, 0, NOW(), NULL, NULL)',
        'INSERT INTO ' . TABLE_COLS . ' VALUES ("Screening/FastQ_files",                  100, 80, 0, 1, 0, "FastQ Files",                          "", "FastQ Files", "FastQ Files", "TEXT", "FastQ Files||text|8", "", "", 1, 1, 1, 0, NOW(), NULL, NULL)',
        'INSERT INTO ' . TABLE_COLS . ' VALUES ("Screening/Variants_file",                100, 80, 0, 1, 0, "Variants File",                        "", "Variants File", "Variants File", "TEXT", "Variants File||text|8", "", "", 1, 1, 1, 0, NOW(), NULL, NULL)',
        'INSERT INTO ' . TABLE_COLS . ' VALUES ("Screening/Pedigree_file",                100, 80, 0, 1, 0, "Pedigree File",                        "", "Details about the parentage of the pro band.", "Pedigree File", "VARCHAR(100)", "Pedigree File||text|8", "", "", 1, 1, 1, 0, NOW(), NULL, NULL)',
        'INSERT INTO ' . TABLE_COLS . ' VALUES ("Screening/Variant_call_group",           100, 80, 0, 1, 0, "Variant Call Group",                   "", "Variant Call Group", "Variant Call Group", "VARCHAR(100)", "Variant Call Group||text|8", "", "", 1, 1, 1, 0, NOW(), NULL, NULL)',
        'INSERT INTO ' . TABLE_COLS . ' VALUES ("Screening/Capture_date",                 100, 80, 0, 1, 0, "Capture Date",                         "", "Capture Date", "Capture Date", "VARCHAR(255)", "Capture Date||text|8", "", "", 1, 1, 1, 0, NOW(), NULL, NULL)',
        'INSERT INTO ' . TABLE_COLS . ' VALUES ("Screening/Sequencing_date",              100, 80, 0, 1, 0, "Sequencing Date",                      "", "The date the sequencing process was complete.", "The date the sequencing process was complete.", "VARCHAR(255)", "Sequencing Date||text|8", "", "", 1, 1, 1, 0, NOW(), NULL, NULL)',
        'INSERT INTO ' . TABLE_COLS . ' VALUES ("Screening/Mean_coverage",                100, 80, 0, 1, 0, "Mean Coverage",                        "", "Mean Coverage", "Mean Coverage", "VARCHAR(10)", "Mean Coverage||text|8", "", "", 1, 1, 1, 0, NOW(), NULL, NULL)',
        'INSERT INTO ' . TABLE_COLS . ' VALUES ("Screening/Duplicate_percentage",         100, 80, 0, 1, 0, "Duplicate Percentage",                 "", "Duplicate Percentage", "Duplicate Percentage", "VARCHAR(10)", "Duplicate Percentage||text|8", "", "", 1, 0, 1, 0, NOW(), NULL, NULL)',
        'INSERT INTO ' . TABLE_COLS . ' VALUES ("Screening/Machine_ID",                   100, 80, 0, 1, 0, "Machine ID",                           "", "Machine ID", "Machine ID", "VARCHAR(20)", "Machine ID||text|8", "", "", 1, 1, 1, 0, NOW(), NULL, NULL)',
        'INSERT INTO ' . TABLE_COLS . ' VALUES ("Screening/DNA_extraction_lab",           100, 80, 0, 1, 0, "DNA Extraction Lab",                   "", "The laboratory who extracted the DNA.", "The laboratory who extracted the DNA", "VARCHAR(20)", "DNA Extraction Lab||text|8", "", "", 1, 1, 1, 0, NOW(), NULL, NULL)',
        'INSERT INTO ' . TABLE_COLS . ' VALUES ("Screening/Sequencing_lab",               100, 80, 0, 1, 0, "Sequencing Lab",                       "", "The laboratory who performed the sequencing.", "The laboratory who performed the sequencing", "VARCHAR(20)", "Sequencing Lab||text|8", "", "", 1, 1, 1, 0, NOW(), NULL, NULL)',
        'INSERT INTO ' . TABLE_COLS . ' VALUES ("Screening/Exome_capture",                100, 80, 0, 1, 0, "Exome Capture",                        "", "Exome Capture", "Exome Capture", "VARCHAR(50)", "Exome Capture||text|8", "", "", 1, 1, 1, 0, NOW(), NULL, NULL)',
        'INSERT INTO ' . TABLE_COLS . ' VALUES ("Screening/Library_preparation",          100, 80, 0, 1, 0, "Library Preparation",                  "", "Library Preparation", "Library Preparation", "VARCHAR(50)", "Library Preparation||text|8", "", "", 1, 1, 1, 0, NOW(), NULL, NULL)',
        'INSERT INTO ' . TABLE_COLS . ' VALUES ("Screening/Barcode_pool_size",            100, 80, 0, 1, 0, "Barcode Pool Size",                    "", "Barcode Pool Size", "Barcode Pool Size", "SMALLINT(3)", "Barcode Pool Size||text|8", "", "", 1, 1, 1, 0, NOW(), NULL, NULL)',
        'INSERT INTO ' . TABLE_COLS . ' VALUES ("Screening/Read_type",                    100, 80, 0, 1, 0, "Read Type",                            "", "Read Type", "Read Type", "VARCHAR(10)", "Read Type||text|8", "", "", 1, 1, 1, 0, NOW(), NULL, NULL)',
        'INSERT INTO ' . TABLE_COLS . ' VALUES ("Screening/Machine_type",                 100, 80, 0, 1, 0, "Machine Type",                         "", "Machine Type", "Machine Type", "VARCHAR(50)", "Machine Type||text|8", "", "", 1, 1, 1, 0, NOW(), NULL, NULL)',
        'INSERT INTO ' . TABLE_COLS . ' VALUES ("Screening/Sequencing_chemistry",         100, 80, 0, 1, 0, "Sequencing Chemistry",                 "", "Sequencing Chemistry", "Sequencing Chemistry", "VARCHAR(50)", "Sequencing Chemistry||text|8", "", "", 1, 1, 1, 0, NOW(), NULL, NULL)',
        'INSERT INTO ' . TABLE_COLS . ' VALUES ("Screening/Sequencing_software",          100, 80, 0, 1, 0, "Sequencing Software",                  "", "Sequencing Software", "Sequencing Software", "VARCHAR(50)", "Sequencing Software||text|8", "", "", 1, 1, 1, 0, NOW(), NULL, NULL)',
        'INSERT INTO ' . TABLE_COLS . ' VALUES ("Screening/Demultiplex_software",         100, 80, 0, 1, 0, "Demultiplex Software",                 "", "Demultiplex Software", "Demultiplex Software", "VARCHAR(50)", "Demultiplex Software||text|8", "", "", 1, 1, 1, 0, NOW(), NULL, NULL)',
        'INSERT INTO ' . TABLE_COLS . ' VALUES ("Screening/Sequencing_contact",           100, 80, 0, 1, 0, "Sequencing Contact",                   "", "The email address used to contact the sequencing team.", "The email address used to contact the sequencing team.", "VARCHAR(100)", "Sequencing Contact||text|8", "", "", 1, 1, 1, 0, NOW(), NULL, NULL)',
        'INSERT INTO ' . TABLE_COLS . ' VALUES ("Screening/Pipeline_contact",             100, 80, 0, 1, 0, "Pipeline Contact",                     "", "The email address used to contact the pipeline team", "The email address used to contact the pipeline team.", "VARCHAR(100)", "Pipeline Contact||text|8", "", "", 1, 1, 1, 0, NOW(), NULL, NULL)',

        // VARIANT COLUMNS
        //                                      COL ID,                                                                     ORDER,WIDTH,HGVS,STND,MAND,TITLE,                                   FORM DESCRIPTION, SHORT LEGEND,  FULL LEGEND,  MYSQL,  FORM DEFINITION  SELECT OPTIONS, REGEXP,  PUBLIC (VIEW,ADD), COUNT?, CREATED_BY, CREATED DATE, EDITED BY, EDITED DATE.
        'INSERT INTO ' . TABLE_COLS . ' VALUES ("VariantOnGenome/Reference",                                                  6, 200, 1, 1, 0, "Reference",                                     "", "Reference to publication describing the variant.", "Reference to publication describing the variant, including links to OMIM (when available), PubMed or or other source, e.g. \"den Dunnen ASHG2003 P2346\".", "VARCHAR(255)", "Reference||text|50", "", "", 1, 1, 1, 0, NOW(), NULL, NULL)',
        'INSERT INTO ' . TABLE_COLS . ' VALUES ("VariantOnGenome/Alamut",                                                   255, 100, 0, 1, 0, "Alamut link",                                   "", "Links to the variant in Alamut.", "Links to the variant in Alamut.", "VARCHAR(255)", "Alamut link||text|20", "", "", 1, 1, 1, 0, NOW(), NULL, NULL)',

        'INSERT INTO ' . TABLE_COLS . ' VALUES ("VariantOnGenome/DBID",                                                       7, 120, 1, 1, 1, "DB-ID",                                         "NOTE: This field will be predicted and filled in by LOVD, if left empty.", "Database ID of variant starting with the HGNC gene symbol, followed by an underscore (_) and a six digit number (e.g. DMD_012345). _000000 is used for variants where DNA was not analysed (change predicted from RNA analysis), variants seen in animal models or variants not seen in humans but functionally tested in vitro.", "Database ID of variant, grouping multiple observations of the same variant together, starting with the HGNC gene symbol, followed by an underscore (_) and a six digit number (e.g. DMD_012345). _000000 is used for variants where DNA was not analysed (change predicted from RNA analysis), variants seen in animal models or variants not seen in humans but functionally tested in vitro.", "VARCHAR(50)", "ID|This ID is used to group multiple observations of the same variant together. This field will be predicted and filled in by LOVD when left empty. The ID starts with the HGNC gene symbol of the transcript most influenced by the variant or otherwise the closest gene, followed by an underscore (_) and a six digit number (e.g. DMD_012345). _000000 is used for variants where DNA was not analysed (change predicted from RNA analysis), variants seen in animal models or variants not seen in humans but functionally tested in vitro.|text|20", "", "/^(chr(\\\\d{1,2}|[XYM])|(C(\\\\d{1,2}|[XYM])orf[\\\\d][\\\\dA-Z]*-|[A-Z][A-Z0-9]+-)?(C(\\\\d{1,2}|[XYM])orf[\\\\d][\\\\dA-Z]*|[A-Z][A-Z0-9-]+))_\\\\d{6}$/", 1, 0, 1, 0, NOW(), NULL, NULL)',
        'INSERT INTO ' . TABLE_COLS . ' VALUES ("VariantOnGenome/dbSNP",                                                      8, 120, 0, 1, 0, "dbSNP ID",                                      "", "The dbSNP ID.", "The dbSNP ID.", "VARCHAR(15)", "dbSNP ID|If available, please fill in the dbSNP ID, such as rs12345678.|text|10", "", "/^[rs]s\\\\d+$/", 0, 0, 1, 0, NOW(), NULL, NULL)',
        'INSERT INTO ' . TABLE_COLS . ' VALUES ("VariantOnGenome/DNA",                                                        2, 200, 1, 1, 1, "DNA change (genomic)",                          "", "Description of variant at DNA level, based on the genomic DNA reference sequence (following HGVS recommendations).", "Description of variant at DNA level, based on the genomic DNA reference sequence (following HGVS recommendations).<BR>\r\n<UL style=\"margin-top : 0px;\">\r\n  <LI>g.12345678C>T</LI>\r\n  <LI>g.12345678_12345890del</LI>\r\n  <LI>g.12345678_12345890dup</LI>\r\n</UL>", "VARCHAR(150)", "Genomic DNA change (HGVS format)|Description of variant at DNA level, based on the genomic DNA reference sequence (following HGVS recommendations); e.g. g.12345678C>T, g.12345678_12345890del, g.12345678_12345890dup.|text|30", "", "", 1, 1, 1, 0, NOW(), NULL, NULL)',
        'INSERT INTO ' . TABLE_COLS . ' VALUES ("VariantOnGenome/Sequencing/Quality",                                       255, 100, 0, 1, 0, "Sequencing quality",                            "", "Sequencing quality.", "Sequencing quality, as presented in the input file.", "DECIMAL(9,2)", "Sequencing quality||text|6", "", "", 0, 0, 1, 0, NOW(), NULL, NULL)',
        'INSERT INTO ' . TABLE_COLS . ' VALUES ("VariantOnGenome/Sequencing/Filter",                                        255, 100, 0, 1, 0, "Sequencing filter",                             "", "Sequencing filter.", "Sequencing filter output, as presented in the input file.", "VARCHAR(50)", "Sequencing filter||text|40", "", "", 0, 0, 1, 0, NOW(), NULL, NULL)',
        'INSERT INTO ' . TABLE_COLS . ' VALUES ("VariantOnGenome/Sequencing/Depth/Unfiltered_All",                          255, 100, 0, 1, 0, "Unfiltered Read Depth",                         "", "Unfiltered Read Depth", "Unfiltered depth of coverage across all samples. (reads with MQ=255 or with bad mates are filtered). .", "VARCHAR(255)", "Unfiltered Read Depth||text|6", "", "", 0, 0, 1, 0, NOW(), NULL, NULL)',
        'INSERT INTO ' . TABLE_COLS . ' VALUES ("VariantOnGenome/Sequencing/Depth/Total",                                   255, 100, 0, 1, 0, "Read depth",                                    "", "Sequencing read depth.", "Sequencing read depth, as presented in the input file. Filtered for each sample", "SMALLINT UNSIGNED", "Read depth||text|6", "", "", 0, 0, 1, 0, NOW(), NULL, NULL)',
        'INSERT INTO ' . TABLE_COLS . ' VALUES ("VariantOnGenome/Sequencing/Depth/Alt",                                     255, 100, 0, 1, 0, "Read depth Alt",                                "", "Sequencing read depth of Alt allele.", "Sequencing read depth of Alt allele, as presented in the input file.", "SMALLINT UNSIGNED", "Read depth Alt||text|6", "", "", 0, 0, 1, 0, NOW(), NULL, NULL)',
        'INSERT INTO ' . TABLE_COLS . ' VALUES ("VariantOnGenome/Sequencing/Depth/Alt/Fraction",                            255, 100, 0, 1, 0, "Read depth Alt (fraction)",                     "", "Sequencing read depth of Alt allele as a fraction.", "Sequencing read depth of Alt allele, as presented in the input file, as the fraction of the total number of reads of Ref and Alt allele together.", "DECIMAL(3,2)", "Read depth Alt (fraction)||text|6", "", "", 0, 0, 1, 0, NOW(), NULL, NULL)',
        'INSERT INTO ' . TABLE_COLS . ' VALUES ("VariantOnGenome/Sequencing/Depth/Ref",                                     255, 100, 0, 1, 0, "Read depth Ref",                                "", "Sequencing read depth of Ref allele.", "Sequencing read depth of Ref allele, as presented in the input file.", "SMALLINT UNSIGNED", "Read depth Ref||text|6", "", "", 0, 0, 1, 0, NOW(), NULL, NULL)',
        'INSERT INTO ' . TABLE_COLS . ' VALUES ("VariantOnGenome/Sequencing/Allele_Balance",                                255, 100, 0, 1, 0, "Allele Balance",                                "", "Allele Balance.", "Allele balance for each het genotype.", "DECIMAL(9,2)", "Allele balance||text|6", "", "", 0, 0, 1, 0, NOW(), NULL, NULL)',
        'INSERT INTO ' . TABLE_COLS . ' VALUES ("VariantOnGenome/Sequencing/Genotype/Quality",                              255, 100, 0, 1, 0, "Genotype Quality",                              "", "Genotype quality.", "Genotype quality.", "DECIMAL(9,2)", "Genotype quality||text|6", "", "", 0, 0, 1, 0, NOW(), NULL, NULL)',
        'INSERT INTO ' . TABLE_COLS . ' VALUES ("VariantOnGenome/Sequencing/Physical_Phasing_ID",                           255, 100, 0, 1, 0, "Physical Phasing ID",                           "", "Physical phasing ID information", "Physical phasing ID information, where each unique ID within a given sample (but not across samples) connects records within a phasing group.", "VARCHAR(255)", "Physical Phasing ID||text|6", "", "", 0, 0, 1, 0, NOW(), NULL, NULL)',
        'INSERT INTO ' . TABLE_COLS . ' VALUES ("VariantOnGenome/Sequencing/Phredscaled_Likelihoods",                       255, 100, 0, 1, 0, "Phredscaled Likelihoods",                       "", "Normalized, Phred-scaled likelihoods for genotypes", "Normalized, Phred-scaled likelihoods for genotypes.", "VARCHAR(255)", "Phredscaled Likelihoods||text|6", "", "", 0, 0, 1, 0, NOW(), NULL, NULL)',
        'INSERT INTO ' . TABLE_COLS . ' VALUES ("VariantOnGenome/Sequencing/Allele/Balance_Het",                            255, 100, 0, 1, 0, "Allele Balance Het",                            "", "Allele Balance Heterozygous", "Allele Balance for heterozygous calls (ref/(ref+alt)).", "VARCHAR(255)", "Allele Balance Heterozygous||text|6", "", "", 0, 0, 1, 0, NOW(), NULL, NULL)',
        'INSERT INTO ' . TABLE_COLS . ' VALUES ("VariantOnGenome/Sequencing/Allele/Balance_Homo",                           255, 100, 0, 1, 0, "Allele Balance Homo",                           "", "Allele Balance Homozygous", "Allele Balance for homozygous calls (A/(A+O)) where A is the allele (ref or alt) and O is anything other.", "VARCHAR(255)", "Allele Balance Homozygous||text|6", "", "", 0, 0, 1, 0, NOW(), NULL, NULL)',
        'INSERT INTO ' . TABLE_COLS . ' VALUES ("VariantOnGenome/Sequencing/Allele/Count",                                  255, 100, 0, 1, 0, "ALT Allele Count",                              "", "ALT Allele Count", "Allele count in genotypes, for each ALT allele, in the same order as listed.", "VARCHAR(255)", "ALT Allele Count||text|6", "", "", 0, 0, 1, 0, NOW(), NULL, NULL)',
        'INSERT INTO ' . TABLE_COLS . ' VALUES ("VariantOnGenome/Sequencing/Allele/Frequency",                              255, 100, 0, 1, 0, "ALT Allele Frequency",                          "", "ALT Allele Frequency", "Allele Frequency, for each ALT allele, in the same order as listed.", "VARCHAR(255)", "ALT Allele Frequency||text|6", "", "", 0, 0, 1, 0, NOW(), NULL, NULL)',
        'INSERT INTO ' . TABLE_COLS . ' VALUES ("VariantOnGenome/Sequencing/Allele/Total",                                  255, 100, 0, 1, 0, "Total Alleles",                                 "", "Total number of alleles", "Total number of alleles in called genotypes.", "VARCHAR(255)", "Total number of alleles||text|6", "", "", 0, 0, 1, 0, NOW(), NULL, NULL)',
        'INSERT INTO ' . TABLE_COLS . ' VALUES ("VariantOnGenome/Sequencing/Base_Qualities_Score",                          255, 100, 0, 1, 0, "Base Qualities Score",                          "", "Base Qualities Score", "Z-score from Wilcoxon rank sum test of Alt Vs. Ref base qualities.", "VARCHAR(255)", "Base Qualities Score||text|6", "", "", 0, 0, 1, 0, NOW(), NULL, NULL)',
        'INSERT INTO ' . TABLE_COLS . ' VALUES ("VariantOnGenome/Sequencing/dbSNP_Membership",                              255, 100, 0, 1, 0, "dbSNP Membership",                              "", "dbSNP Membership", "dbSNP Membership.", "VARCHAR(255)", "dbSNP Membership||text|6", "", "", 0, 0, 1, 0, NOW(), NULL, NULL)',
        'INSERT INTO ' . TABLE_COLS . ' VALUES ("VariantOnGenome/Sequencing/Excess_Heterozygosity",                         255, 100, 0, 1, 0, "Excess Heterozygosity",                         "", "Excess Heterozygosity", "Phred-scaled p-value for exact test of excess heterozygosity.", "VARCHAR(255)", "Excess Heterozygosity||text|6", "", "", 0, 0, 1, 0, NOW(), NULL, NULL)',
        'INSERT INTO ' . TABLE_COLS . ' VALUES ("VariantOnGenome/Sequencing/Fisher_Strand_Bias",                            255, 100, 0, 1, 0, "Fisher\'s Strand Bias",                         "", "Fisher\'s Strand Bias", "Phred-scaled p-value using Fisher\'s exact test to detect strand bias.", "VARCHAR(255)", "Fisher\'s Strand Bias||text|6", "", "", 0, 0, 1, 0, NOW(), NULL, NULL)',
        'INSERT INTO ' . TABLE_COLS . ' VALUES ("VariantOnGenome/Sequencing/Genotype/Quality/Mean",                         255, 100, 0, 1, 0, "Mean Genotype Quality",                         "", "Mean of all GQ values", "Mean of all GQ values.", "VARCHAR(255)", "Mean of all GQ values||text|6", "", "", 0, 0, 1, 0, NOW(), NULL, NULL)',
        'INSERT INTO ' . TABLE_COLS . ' VALUES ("VariantOnGenome/Sequencing/Haplotype_Likelihood_Score",                    255, 100, 0, 1, 0, "Haplotype Likelihood Score",                    "", "Haplotype Likelihood Score", "Z-score from Wilcoxon rank sum test of Alt Vs. Ref haplotype likelihoods.", "VARCHAR(255)", "Haplotype Likelihood Score||text|6", "", "", 0, 0, 1, 0, NOW(), NULL, NULL)',
        'INSERT INTO ' . TABLE_COLS . ' VALUES ("VariantOnGenome/Sequencing/Max_Likelihood_Exp_Allele_Count",               255, 100, 0, 1, 0, "Max Likelihood Allele Count",                   "", "Maximum Likelihood Expectation Allele Count", "Maximum likelihood expectation (MLE) for the allele counts (not necessarily the same as the AC), for each ALT allele, in the same order as listed.", "VARCHAR(255)", "Maximum Likelihood Expectation Allele Count||text|6", "", "", 0, 0, 1, 0, NOW(), NULL, NULL)',
        'INSERT INTO ' . TABLE_COLS . ' VALUES ("VariantOnGenome/Sequencing/Max_Likelihood_Exp_Allele_Freq",                255, 100, 0, 1, 0, "Max Likelihood Allele Count",                   "", "Maximum Likelihood Expectation Allele Frequency", "Maximum likelihood expectation (MLE) for the allele frequency (not necessarily the same as the AF), for each ALT allele, in the same order as listed.", "VARCHAR(255)", "Maximum Likelihood Expectation Allele Frequency||text|6", "", "", 0, 0, 1, 0, NOW(), NULL, NULL)',
        'INSERT INTO ' . TABLE_COLS . ' VALUES ("VariantOnGenome/Sequencing/Mapping_Quality",                               255, 100, 0, 1, 0, "RMS Mapping Quality",                           "", "RMS Mapping Quality", "RMS Mapping Quality.", "VARCHAR(255)", "RMS Mapping Quality||text|6", "", "", 0, 0, 1, 0, NOW(), NULL, NULL)',
        'INSERT INTO ' . TABLE_COLS . ' VALUES ("VariantOnGenome/Sequencing/Mapping_Quality_Score",                         255, 100, 0, 1, 0, "RMS Mapping Quality Score",                     "", "RMS Mapping Quality Score", "Z-score From Wilcoxon rank sum test of Alt vs. Ref read mapping qualities.", "VARCHAR(255)", "RMS Mapping Quality Score||text|6", "", "", 0, 0, 1, 0, NOW(), NULL, NULL)',
        'INSERT INTO ' . TABLE_COLS . ' VALUES ("VariantOnGenome/Sequencing/Non_diploid_Ratio",                             255, 100, 0, 1, 0, "Overall non-diploid ratio",                     "", "Overall non-diploid ratio", "Overall non-diploid ratio (alleles/(alleles+non-alleles)).", "VARCHAR(255)", "Overall non-diploid ratio||text|6", "", "", 0, 0, 1, 0, NOW(), NULL, NULL)',
        'INSERT INTO ' . TABLE_COLS . ' VALUES ("VariantOnGenome/Sequencing/Genotype_Likelihood_Prior",                     255, 100, 0, 1, 0, "Genotype Likelihood Prior",                     "", "Genotype Likelihood Prior", "Genotype Likelihood Prior.", "VARCHAR(255)", "Genotype Likelihood Prior||text|6", "", "", 0, 0, 1, 0, NOW(), NULL, NULL)',
        'INSERT INTO ' . TABLE_COLS . ' VALUES ("VariantOnGenome/Sequencing/Quality_by_depth",                              255, 100, 0, 1, 0, "Quality by Depth",                              "", "Quality by Depth", "Variant Confidence/Quality by Depth.", "VARCHAR(255)", "Quality by Depth||text|6", "", "", 0, 0, 1, 0, NOW(), NULL, NULL)',
        'INSERT INTO ' . TABLE_COLS . ' VALUES ("VariantOnGenome/Sequencing/Read_Position_Bias_Score",                      255, 100, 0, 1, 0, "Read Position Bias Score",                      "", "Read Position Bias Score", "Z-score from Wilcoxon rank sum test of Alt vs. Ref read position bias.", "VARCHAR(255)", "Read Position Bias Score||text|6", "", "", 0, 0, 1, 0, NOW(), NULL, NULL)',
        'INSERT INTO ' . TABLE_COLS . ' VALUES ("VariantOnGenome/Sequencing/Symmetric_Odds_Ratio",                          255, 100, 0, 1, 0, "Symmetric Odds Ratio",                          "", "Symmetric Odds Ratio", "Symmetric Odds Ratio of 2x2 contingency table to detect strand bias.", "VARCHAR(255)", "Symmetric Odds Ratio||text|6", "", "", 0, 0, 1, 0, NOW(), NULL, NULL)',
        'INSERT INTO ' . TABLE_COLS . ' VALUES ("VariantOnGenome/Sequencing/Variant_Type",                                  255, 100, 0, 1, 0, "Variant type description",                      "", "Variant type description", "Variant type description.", "VARCHAR(255)", "Variant type description||text|6", "", "", 0, 0, 1, 0, NOW(), NULL, NULL)',
        'INSERT INTO ' . TABLE_COLS . ' VALUES ("VariantOnGenome/Sequencing/High_Confidence_DeNovo",                        255, 100, 0, 1, 0, "High Confidence De Novo",                       "", "High Confidence De Novo", "High confidence possible de novo mutation (GQ >= 20 for all trio members)=[comma-delimited list of child samples].", "VARCHAR(255)", "High Confidence De Novo||text|6", "", "", 0, 0, 1, 0, NOW(), NULL, NULL)',
        'INSERT INTO ' . TABLE_COLS . ' VALUES ("VariantOnGenome/Sequencing/Low_Confidence_DeNovo",                         255, 100, 0, 1, 0, "Low Confidence De Novo",                        "", "Low Confidence De Novo", "Low confidence possible de novo mutation (GQ >= 10 for child, GQ > 0 for parents)=[comma-delimited list of child samples].", "VARCHAR(255)", "Low Confidence De Novo||text|6", "", "", 0, 0, 1, 0, NOW(), NULL, NULL)',
        'INSERT INTO ' . TABLE_COLS . ' VALUES ("VariantOnGenome/Sequencing/Source_VCF",                                    255, 100, 0, 1, 0, "Source VCF",                                    "", "Source VCF", "Source VCF for the merged record in CombineVariants.", "VARCHAR(255)", "Source VCF||text|6", "", "", 0, 0, 1, 0, NOW(), NULL, NULL)',
        'INSERT INTO ' . TABLE_COLS . ' VALUES ("VariantOnGenome/Sequencing/Phredscaled_Joint_Likelihood",                  255, 100, 0, 1, 0, "Phred-scaled joint likelihood",                 "", "Phred-scaled joint likelihood", "Phred-scaled joint likelihood of the genotype combination (before applying family priors).", "VARCHAR(255)", "Phred-scaled joint likelihood||text|6", "", "", 0, 0, 1, 0, NOW(), NULL, NULL)',
        'INSERT INTO ' . TABLE_COLS . ' VALUES ("VariantOnGenome/Sequencing/Phredscaled_Joint_Probability",                 255, 100, 0, 1, 0, "Phred-scaled joint posterior probability",      "", "Phred-scaled joint posterior probability", "Phred-scaled joint posterior probability of the genotype combination (after applying family priors).", "VARCHAR(255)", "Phred-scaled joint posterior probability||text|6", "", "", 0, 0, 1, 0, NOW(), NULL, NULL)',
        'INSERT INTO ' . TABLE_COLS . ' VALUES ("VariantOnGenome/Sequencing/Phredscaled_Probabilities",                     255, 100, 0, 1, 0, "Phred-scaled Posterior Probabilities",          "", "Phred-scaled Posterior Genotype Probabilities", "Phred-scaled Posterior Genotype Probabilities.", "VARCHAR(255)", "Phred-scaled Posterior Probabilities||text|6", "", "", 0, 0, 1, 0, NOW(), NULL, NULL)',

        'INSERT INTO ' . TABLE_COLS . ' VALUES ("VariantOnGenome/Sequencing/Father/Depth/Total",                            255, 100, 0, 1, 0, "Father\'s read depth",                          "", "Sequencing read depth of unaffected father.", "Sequencing read depth of unaffected father, as presented in the input file. Filtered for each sample.", "SMALLINT UNSIGNED", "Read depth father||text|6", "", "", 0, 0, 1, 0, NOW(), NULL, NULL)',
        'INSERT INTO ' . TABLE_COLS . ' VALUES ("VariantOnGenome/Sequencing/Father/Depth/Alt",                              255, 100, 0, 1, 0, "Father\'s read depth Alt",                      "", "Sequencing read depth of Alt allele of unaffected father.", "Sequencing read depth of Alt allele of unaffected father, as presented in the input file.", "SMALLINT UNSIGNED", "Read depth father Alt||text|6", "", "", 0, 0, 1, 0, NOW(), NULL, NULL)',
        'INSERT INTO ' . TABLE_COLS . ' VALUES ("VariantOnGenome/Sequencing/Father/Depth/Alt/Fraction",                     255, 100, 0, 1, 0, "Father\'s read depth Alt (fraction)",           "", "Sequencing read depth of Alt allele of unaffected father as a fraction.", "Sequencing read depth of Alt allele of unaffected father, as presented in the input file, as the fraction of the total number of reads of Ref and Alt allele together.", "DECIMAL(9,2)", "Read depth father Alt (fraction)||text|6", "", "", 0, 0, 1, 0, NOW(), NULL, NULL)',
        'INSERT INTO ' . TABLE_COLS . ' VALUES ("VariantOnGenome/Sequencing/Father/Depth/Ref",                              255, 100, 0, 1, 0, "Father\'s read depth Ref",                      "", "Sequencing read depth of Ref allele of unaffected father.", "Sequencing read depth of Ref allele, as presented in the input file.", "SMALLINT UNSIGNED", "Read depth father Ref||text|6", "", "", 0, 0, 1, 0, NOW(), NULL, NULL)',
        'INSERT INTO ' . TABLE_COLS . ' VALUES ("VariantOnGenome/Sequencing/Father/GenoType",                               255, 100, 0, 1, 0, "Father\'s genotype",                            "", "The genotype of the unaffected father.", "The genotype of the unaffected father.", "VARCHAR(3)", "Genotype of father||text|4", "", "", 0, 0, 1, 0, NOW(), NULL, NULL)',
        'INSERT INTO ' . TABLE_COLS . ' VALUES ("VariantOnGenome/Sequencing/Father/VarPresent",                             255, 100, 0, 1, 0, "Variant present in father",                     "", "Is the variant present in the unaffected father? Scores 1-6.", "Is the variant present in the unaffected father? Scores 1-6. 1 certainly not present, 6 certainly present.", "TINYINT UNSIGNED", "Variant present in father||select|1|true|false|false", "1=Not present with DP>20\r\n2=Not present with DP 8-20\r\n3=Not present with DP<8\r\n4=Present with <10% of reads\r\n5=Present with >10% of reads\r\n6=Called by pipeline", "", 0, 0, 1, 0, NOW(), NULL, NULL)',
        'INSERT INTO ' . TABLE_COLS . ' VALUES ("VariantOnGenome/Sequencing/Father/Allele_Balance",                         255, 100, 0, 1, 0, "Father\'s Allele Balance",                      "", "Father\'s allele balance.", "Father\'s allele balance.", "DECIMAL(9,2)", "Father\'s allele balance||text|6", "", "", 0, 0, 1, 0, NOW(), NULL, NULL)',
        'INSERT INTO ' . TABLE_COLS . ' VALUES ("VariantOnGenome/Sequencing/Father/Genotype/Quality",                       255, 100, 0, 1, 0, "Father\'s Genotype Quality",                    "", "Father\'s genotype quality", "Father\'s genotype quality.", "DECIMAL(9,2)", "Father\'s genotype quality||text|6", "", "", 0, 0, 1, 0, NOW(), NULL, NULL)',
        'INSERT INTO ' . TABLE_COLS . ' VALUES ("VariantOnGenome/Sequencing/Father/Phredscaled_Likelihoods",                255, 100, 0, 1, 0, "Father\'s Phredscaled Likelihoods",             "", "Father\'s normalized, Phred-scaled likelihoods for genotypes", "Father\'s normalized, Phred-scaled likelihoods for genotypes.", "VARCHAR(255)", "Father\'s Phredscaled Likelihoods||text|6", "", "", 0, 0, 1, 0, NOW(), NULL, NULL)',
        'INSERT INTO ' . TABLE_COLS . ' VALUES ("VariantOnGenome/Sequencing/Father/Physical_Phasing_ID",                    255, 100, 0, 1, 0, "Father\'s Physical Phasing ID",                 "", "Father\'s Physical phasing ID information", "Father\'s physical phasing ID information, where each unique ID within a given sample (but not across samples) connects records within a phasing group.", "VARCHAR(255)", "Father\'s Physical Phasing ID||text|6", "", "", 0, 0, 1, 0, NOW(), NULL, NULL)',
        'INSERT INTO ' . TABLE_COLS . ' VALUES ("VariantOnGenome/Sequencing/Father/Phredscaled_Joint_Likelihood",           255, 100, 0, 1, 0, "Father\'s Phred-scaled Joint Likelihood",       "", "Father\'s Phred-scaled joint likelihood", "Father\'s Phred-scaled joint likelihood of the genotype combination (before applying family priors).", "VARCHAR(255)", "Father\'s Phred-scaled joint likelihood||text|6", "", "", 0, 0, 1, 0, NOW(), NULL, NULL)',
        'INSERT INTO ' . TABLE_COLS . ' VALUES ("VariantOnGenome/Sequencing/Father/Phredscaled_Joint_Probability",          255, 100, 0, 1, 0, "Father\'s Joint Posterior Probability",         "", "Father\'s Phred-scaled joint posterior probability", "Father\'s Phred-scaled joint posterior probability of the genotype combination (after applying family priors).", "VARCHAR(255)", "Father\'s Phred-scaled joint posterior probability||text|6", "", "", 0, 0, 1, 0, NOW(), NULL, NULL)',
        'INSERT INTO ' . TABLE_COLS . ' VALUES ("VariantOnGenome/Sequencing/Father/Phredscaled_Probabilities",              255, 100, 0, 1, 0, "Father\'s Posterior Probabilities",             "", "Father\'s Phred-scaled Posterior Probabilities", "Father\'s Phred-scaled Posterior Genotype Probabilities.", "VARCHAR(255)", "Father\'s Phred-scaled Posterior Probabilities||text|6", "", "", 0, 0, 1, 0, NOW(), NULL, NULL)',

        'INSERT INTO ' . TABLE_COLS . ' VALUES ("VariantOnGenome/Sequencing/Mother/Depth/Total",                            255, 100, 0, 1, 0, "Mother\'s read depth",                          "", "Sequencing read depth of unaffected mother.", "Sequencing read depth of unaffected mother, as presented in the input file. Filtered for each sample.", "SMALLINT UNSIGNED", "Read depth mother||text|6", "", "", 0, 0, 1, 0, NOW(), NULL, NULL)',
        'INSERT INTO ' . TABLE_COLS . ' VALUES ("VariantOnGenome/Sequencing/Mother/Depth/Alt",                              255, 100, 0, 1, 0, "Mother\'s read depth Alt",                      "", "Sequencing read depth of Alt allele of unaffected mother.", "Sequencing read depth of Alt allele of unaffected mother, as presented in the input file.", "SMALLINT UNSIGNED", "Read depth mother Alt||text|6", "", "", 0, 0, 1, 0, NOW(), NULL, NULL)',
        'INSERT INTO ' . TABLE_COLS . ' VALUES ("VariantOnGenome/Sequencing/Mother/Depth/Alt/Fraction",                     255, 100, 0, 1, 0, "Mother\'s read depth Alt (fraction)",           "", "Sequencing read depth of Alt allele of unaffected mother as a fraction.", "Sequencing read depth of Alt allele of unaffected mother, as presented in the input file, as the fraction of the total number of reads of Ref and Alt allele together.", "DECIMAL(9,2)", "Read depth mother Alt (fraction)||text|6", "", "", 0, 0, 1, 0, NOW(), NULL, NULL)',
        'INSERT INTO ' . TABLE_COLS . ' VALUES ("VariantOnGenome/Sequencing/Mother/Depth/Ref",                              255, 100, 0, 1, 0, "Mother\'s read depth Ref",                      "", "Sequencing read depth of Ref allele of unaffected mother.", "Sequencing read depth of Ref allele of unaffected mother, as presented in the input file.", "SMALLINT UNSIGNED", "Read depth mother Ref||text|6", "", "", 0, 0, 1, 0, NOW(), NULL, NULL)',
        'INSERT INTO ' . TABLE_COLS . ' VALUES ("VariantOnGenome/Sequencing/Mother/GenoType",                               255, 100, 0, 1, 0, "Mother\'s genotype",                            "", "The genotype of the unaffected mother.", "The genotype of the unaffected mother.", "VARCHAR(3)", "Genotype of mother||text|4", "", "", 0, 0, 1, 0, NOW(), NULL, NULL)',
        'INSERT INTO ' . TABLE_COLS . ' VALUES ("VariantOnGenome/Sequencing/Mother/VarPresent",                             255, 100, 0, 1, 0, "Variant present in mother",                     "", "Is the variant present in the unaffected mother? Scores 1-6.", "Is the variant present in the unaffected mother? Scores 1-6. 1 certainly not present, 6 certainly present.", "TINYINT UNSIGNED", "Variant present in mother||select|1|true|false|false", "1=Not present with DP>20\r\n2=Not present with DP 8-20\r\n3=Not present with DP<8\r\n4=Present with <10% of reads\r\n5=Present with >10% of reads\r\n6=Called by pipeline", "", 0, 0, 1, 0, NOW(), NULL, NULL)',
        'INSERT INTO ' . TABLE_COLS . ' VALUES ("VariantOnGenome/Sequencing/Mother/Allele_Balance",                         255, 100, 0, 1, 0, "Mother\'s Allele Balance",                      "", "Mother\'s allele balance", "Mother\'s allele balance.", "DECIMAL(9,2)", "Mother\'s allele balance||text|6", "", "", 0, 0, 1, 0, NOW(), NULL, NULL)',
        'INSERT INTO ' . TABLE_COLS . ' VALUES ("VariantOnGenome/Sequencing/Mother/Genotype/Quality",                       255, 100, 0, 1, 0, "Mother\'s Genotype Quality",                    "", "Mother\'s genotype quality", "Mother\'s genotype quality.", "DECIMAL(9,2)", "Mother\'s genotype quality||text|6", "", "", 0, 0, 1, 0, NOW(), NULL, NULL)',
        'INSERT INTO ' . TABLE_COLS . ' VALUES ("VariantOnGenome/Sequencing/Mother/Phredscaled_Likelihoods",                255, 100, 0, 1, 0, "Mother\'s Phredscaled Likelihoods",             "", "Mother\'s normalized, Phred-scaled likelihoods for genotypes", "Mother\'s normalized, Phred-scaled likelihoods for genotypes.", "VARCHAR(255)", "Father\'s Phredscaled Likelihoods||text|6", "", "", 0, 0, 1, 0, NOW(), NULL, NULL)',
        'INSERT INTO ' . TABLE_COLS . ' VALUES ("VariantOnGenome/Sequencing/Mother/Physical_Phasing_ID",                    255, 100, 0, 1, 0, "Mother\'s Physical Phasing ID",                 "", "Mother\'s Physical phasing ID information", "Mother\'s physical phasing ID information, where each unique ID within a given sample (but not across samples) connects records within a phasing group.", "VARCHAR(255)", "Mother\'s Physical Phasing ID||text|6", "", "", 0, 0, 1, 0, NOW(), NULL, NULL)',
        'INSERT INTO ' . TABLE_COLS . ' VALUES ("VariantOnGenome/Sequencing/Mother/Phredscaled_Joint_Likelihood",           255, 100, 0, 1, 0, "Mother\'s Phred-scaled Joint Likelihood",       "", "Mother\'s Phred-scaled joint likelihood", "Mother\'s Phred-scaled joint likelihood of the genotype combination (before applying family priors).", "VARCHAR(255)", "Mother\'s Phred-scaled joint likelihood||text|6", "", "", 0, 0, 1, 0, NOW(), NULL, NULL)',
        'INSERT INTO ' . TABLE_COLS . ' VALUES ("VariantOnGenome/Sequencing/Mother/Phredscaled_Joint_Probability",          255, 100, 0, 1, 0, "Mother\'s Joint Posterior Probability",         "", "Mother\'s Phred-scaled joint posterior probability", "Mother\'s Phred-scaled joint posterior probability of the genotype combination (after applying family priors).", "VARCHAR(255)", "Mother\'s joint posterior probability||text|6", "", "", 0, 0, 1, 0, NOW(), NULL, NULL)',
        'INSERT INTO ' . TABLE_COLS . ' VALUES ("VariantOnGenome/Sequencing/Mother/Phredscaled_Probabilities",              255, 100, 0, 1, 0, "Mother\'s Posterior Probabilities",             "", "Mother\'s Phred-scaled Posterior Probabilities", "Mother\'s Phred-scaled Posterior Genotype Probabilities.", "VARCHAR(255)", "Mother\'s Phred-scaled Posterior Probabilities||text|6", "", "", 0, 0, 1, 0, NOW(), NULL, NULL)',

        // VARIANT ON TRANSCRIPT
        'INSERT INTO ' . TABLE_COLS . ' VALUES ("VariantOnGenome/Frequency/1000G",                                          255, 100, 0, 1, 0, "gMAF 1000g",                                    "", "Allele frequency from the 1000 Genomes project.", "Allele frequency from the 1000 Genomes project.", "DECIMAL(5,4)", "1000G allele frequency||text|6", "", "", 0, 0, 1, 0, NOW(), NULL, NULL)',
        'INSERT INTO ' . TABLE_COLS . ' VALUES ("VariantOnGenome/Frequency/ExAC",                                           255, 100, 0, 1, 0, "ExAC",                                          "", "Frequency of existing variant in ExAC combined population", "Frequency of existing variant in ExAC combined population.", "FLOAT UNSIGNED", "ExAC frequency||text|6", "", "", 0, 0, 1, 0, NOW(), NULL, NULL)',
        'INSERT INTO ' . TABLE_COLS . ' VALUES ("VariantOnGenome/Frequency/EVS",                                            255, 100, 0, 1, 0, "EVS",                                           "", "Allele frequency from the EVS project.", "Allele frequency from the EVS project.", "DECIMAL(5,4)", "EVS allele frequency||text|6", "", "", 0, 0, 1, 0, NOW(), NULL, NULL)',

        // REQUIRED TO IMPORT DATA
        'INSERT INTO ' . TABLE_COLS . ' VALUES ("VariantOnTranscript/RNA",                                                    6, 200, 1, 1, 1, "RNA change",                                    "", "Description of variant at RNA level (following HGVS recommendations).", "Description of variant at RNA level (following HGVS recommendations).<BR>\r\n<UL style=\"margin-top : 0px;\">\r\n  <LI>r.123c>u</LI>\r\n  <LI>r.? = unknown</LI>\r\n  <LI>r.(?) = RNA not analysed but probably transcribed copy of DNA variant</LI>\r\n  <LI>r.spl? = RNA not analysed but variant probably affects splicing</LI>\r\n  <LI>r.(spl?) = RNA not analysed but variant may affect splicing</LI>\r\n  <LI>r.0? = change expected to abolish transcription</LI>\r\n</UL>", "VARCHAR(100)", "RNA change (HGVS format)|Description of variant at RNA level (following HGVS recommendations); e.g. r.123c>u, r.? = unknown, r.(?) = RNA not analysed but probably transcribed copy of DNA variant, r.spl? = RNA not analysed but variant probably affects splicing, r.(spl?) = RNA not analysed but variant may affect splicing.|text|30", "", "", 1, 1, 1, 0, NOW(), NULL, NULL)',
        'INSERT INTO ' . TABLE_COLS . ' VALUES ("VariantOnTranscript/Distance_to_splice_site",                              10, 150, 0, 1, 0, "Splice distance",                                "", "The distance to the nearest splice site.", "The distance to the nearest splice site.", "MEDIUMINT(8) UNSIGNED", "Distance to splice site||text|8", "", "", 0, 0, 1, 0, NOW(), NULL, NULL)',

        'INSERT INTO ' . TABLE_COLS . ' VALUES ("VariantOnTranscript/Consequence_Variant_Allele",                           255, 100, 0, 1, 0, "Consequence Variant Allele",                    "", "Consequence Variant Allele", "the variant allele used to calculate the consequence.", "VARCHAR(200)", "Consequence Variant Allele||text|6", "", "", 0, 0, 1, 0, NOW(), NULL, NULL)',
        'INSERT INTO ' . TABLE_COLS . ' VALUES ("VariantOnTranscript/Consequence_Type",                                     255, 100, 0, 1, 0, "Consequence Type",                              "", "Consequence Type", "Consequence type of this variant.", "VARCHAR(200)", "Consequence Type||text|6", "", "", 0, 0, 1, 0, NOW(), NULL, NULL)',
        'INSERT INTO ' . TABLE_COLS . ' VALUES ("VariantOnTranscript/Consequence_Impact",                                   255, 100, 0, 1, 0, "Consequence Impact",                            "", "Consequence Impact", "the impact modifier for the consequence type.", "VARCHAR(200)", "Consequence Impact||text|6", "", "", 0, 0, 1, 0, NOW(), NULL, NULL)',
        //'INSERT INTO ' . TABLE_COLS . ' VALUES ("VariantOnTranscript/Sequencing/Gene_Symbol",                               255, 100, 0, 1, 0, "Gene Symbol",                                   "", "Gene Symbol", "The gene Symbol.", "VARCHAR(200)", "Gene Symbol||text|6", "", "", 0, 0, 1, 0, NOW(), NULL, NULL)',
        'INSERT INTO ' . TABLE_COLS . ' VALUES ("VariantOnTranscript/Emsembl_Stable_ID",                                    255, 100, 0, 1, 0, "Ensembl Stable ID",                             "", "Ensembl Stable ID", "Ensembl stable ID of affected gene.", "VARCHAR(200)", "Ensembl Stable ID||text|6", "", "", 0, 0, 1, 0, NOW(), NULL, NULL)',
        'INSERT INTO ' . TABLE_COLS . ' VALUES ("VariantOnTranscript/Feature_Type",                                         255, 100, 0, 1, 0, "Feature Type",                                  "", "Feature Type", "type of feature. Currently one of Transcript, RegulatoryFeature, MotifFeature..", "VARCHAR(200)", "Feature Type||text|6", "", "", 0, 0, 1, 0, NOW(), NULL, NULL)',
        'INSERT INTO ' . TABLE_COLS . ' VALUES ("VariantOnTranscript/Biotype",                                              255, 100, 0, 1, 0, "Biotype",                                       "", "Biotype", "Biotype of transcript or regulatory feature.", "VARCHAR(200)", "Biotype||text|6", "", "", 0, 0, 1, 0, NOW(), NULL, NULL)',
        'INSERT INTO ' . TABLE_COLS . ' VALUES ("VariantOnTranscript/Exon",                                                 255, 100, 0, 1, 0, "Exon",                                          "", "Exon", "the exon number (out of total number).", "VARCHAR(200)", "Exon||text|6", "", "", 0, 0, 1, 0, NOW(), NULL, NULL)',
        'INSERT INTO ' . TABLE_COLS . ' VALUES ("VariantOnTranscript/Intron",                                               255, 100, 0, 1, 0, "Intron",                                        "", "Intron", "the intron number (out of total number).", "VARCHAR(200)", "Intron||text|6", "", "", 0, 0, 1, 0, NOW(), NULL, NULL)',
        'INSERT INTO ' . TABLE_COLS . ' VALUES ("VariantOnTranscript/DNA",                                                    3, 200, 1, 1, 1, "DNA change (cDNA)",                             "", "Description of variant at DNA level, based on a coding DNA reference sequence (following HGVS recommendations).", "Description of variant at DNA level, based on a coding DNA reference sequence (following HGVS recommendations); e.g. c.123C>T, c.123_145del, c.123_126dup.", "VARCHAR(100)", "DNA change (HGVS format)|Description of variant at DNA level, based on a coding DNA reference sequence (following HGVS recommendations); e.g. c.123C>T, c.123_145del, c.123_126dup.|text|30", "", "", 1, 1, 1, 0, NOW(), NULL, NULL)',
        'INSERT INTO ' . TABLE_COLS . ' VALUES ("VariantOnTranscript/Protein",                                                7, 200, 1, 1, 1, "Protein",                                       "", "Description of variant at protein level (following HGVS recommendations).", "Description of variant at protein level (following HGVS recommendations).<BR>\r\n<UL style=\"margin-top : 0px;\">\r\n  <LI>p.(Arg345Pro) = change predicted from DNA (RNA not analysed)</LI>\r\n  <LI>p.Arg345Pro = change derived from RNA analysis</LI>\r\n  <LI>p.? = unknown effect</LI>\r\n  <LI>p.0? = probably no protein produced</LI>\r\n</UL>", "VARCHAR(100)", "Protein change (HGVS format)|Description of variant at protein level (following HGVS recommendations); e.g. p.(Arg345Pro) = change predicted from DNA (RNA not analysed), p.Arg345Pro = change derived from RNA analysis, p.0 (no protein produced), p.? (unknown effect).|text|30", "", "", 1, 1, 1, 0, NOW(), NULL, NULL)',
        'INSERT INTO ' . TABLE_COLS . ' VALUES ("VariantOnTranscript/cDNA_Position",                                        255, 100, 0, 1, 0, "cDNA Position",                                 "", "cDNA Position", "relative position of base pair in cDNA sequence.", "VARCHAR(200)", "cDNA Position||text|6", "", "", 0, 0, 1, 0, NOW(), NULL, NULL)',
        'INSERT INTO ' . TABLE_COLS . ' VALUES ("VariantOnTranscript/Position",                                               5, 100, 0, 1, 0, "Position",                                      "", "Position of variant in coding DNA sequence.", "Position of variant in coding DNA sequence; note that coding DNA position can also be derived from the variant description.", "MEDIUMINT(6)", "cDNA position|Note that cDNA position can be derived from the variant description.|text|5", "", "", 1, 1, 1, 0, NOW(), NULL, NULL)',
        'INSERT INTO ' . TABLE_COLS . ' VALUES ("VariantOnTranscript/Protein_Position",                                     255, 100, 0, 1, 0, "Protein Position",                              "", "short", "relative position of amino acid in protein.", "VARCHAR(200)", "short||text|6", "", "", 0, 0, 1, 0, NOW(), NULL, NULL)',
        'INSERT INTO ' . TABLE_COLS . ' VALUES ("VariantOnTranscript/Amino_Acids",                                          255, 100, 0, 1, 0, "Amino Acids",                                   "", "Amino Acids", "only given if the variant affects the protein-coding sequence.", "VARCHAR(200)", "short||text|6", "", "", 0, 0, 1, 0, NOW(), NULL, NULL)',
        'INSERT INTO ' . TABLE_COLS . ' VALUES ("VariantOnTranscript/Alternative_Codons",                                   255, 100, 0, 1, 0, "Alternative Codons",                            "", "Alternative Codons", "the alternative codons with the variant base in upper case.", "VARCHAR(200)", "short||text|6", "", "", 0, 0, 1, 0, NOW(), NULL, NULL)',
       // 'INSERT INTO ' . TABLE_COLS . ' VALUES ("VariantOnTranscript/Sequencing/Existing_Variation",                        255, 100, 0, 1, 0, "Existing Variation",                            "", "Existing Variation", "known identifier of existing variant.", "VARCHAR(200)", "Existing Variation||text|6", "", "", 0, 0, 1, 0, NOW(), NULL, NULL)',
        'INSERT INTO ' . TABLE_COLS . ' VALUES ("VariantOnTranscript/DNA_Strand",                                           255, 100, 0, 1, 0, "DNA Strand",                                    "", "DNA Strand", "the DNA strand (1 or -1) on which the transcript/feature lies.", "VARCHAR(200)", "DNA Strand||text|6", "", "", 0, 0, 1, 0, NOW(), NULL, NULL)',
        'INSERT INTO ' . TABLE_COLS . ' VALUES ("VariantOnTranscript/Canonical_Transcript",                                 255, 100, 0, 1, 0, "Canonical Transcript",                          "", "Canonical Transcript", "a flag indicating if the transcript is denoted as the canonical transcript for this gene.", "VARCHAR(200)", "Canonical Transcript||text|6", "", "", 0, 0, 1, 0, NOW(), NULL, NULL)',
        'INSERT INTO ' . TABLE_COLS . ' VALUES ("VariantOnTranscript/Embsembl_Protein_Identifier",                          255, 100, 0, 1, 0, "Embsembl Protein Identifier",                   "", "Embsembl Protein Identifier", "the Ensembl protein identifier of the affected transcript.", "VARCHAR(200)", "Embsembl Protein Identifier||text|6", "", "", 0, 0, 1, 0, NOW(), NULL, NULL)',
        'INSERT INTO ' . TABLE_COLS . ' VALUES ("VariantOnTranscript/Prediction/SIFT_VEP",                                  255, 100, 0, 1, 0, "SIFT prediction (VEP)",                         "", "SIFT prediction (VEP).", "Predicting the effect of the DNA variant on the function of the protein.", "VARCHAR(50)", "SIFT prediction (VEP)||text|10", "", "", 0, 0, 1, 0, NOW(), NULL, NULL)',
        'INSERT INTO ' . TABLE_COLS . ' VALUES ("VariantOnTranscript/Prediction/SIFT_Score_VEP",                            255, 100, 0, 1, 0, "SIFT score (VEP)",                              "", "SIFT score (VEP).", "The SIFT score, predicting the effect of the DNA variant on the function of the protein.", "DECIMAL(3,2)", "SIFT score (VEP)||text|6", "", "", 0, 0, 1, 0, NOW(), NULL, NULL)',
        'INSERT INTO ' . TABLE_COLS . ' VALUES ("VariantOnTranscript/Prediction/PolyPhen_VEP",                              8, 200, 0, 1, 0,   "PolyPhen (VEP)",                                "", "PolyPhen Prediction (VEP).", "Effect of variant, predicted by PolyPhen.", "VARCHAR(50)", "PolyPhen prediction (VEP)||text|5", "", "", 0, 0, 1, 0, NOW(), NULL, NULL)',
        'INSERT INTO ' . TABLE_COLS . ' VALUES ("VariantOnTranscript/Prediction/PolyPhen_Score_VEP",                        255, 100, 0, 1, 0, "PolyPhen score (VEP)",                          "", "PolyPhen Score (VEP).", "Effect of variant, predicted by PolyPhen as a score.", "DECIMAL(4,3)", "PolyPhen score (VEP)||text|6", "", "", 0, 0, 1, 0, NOW(), NULL, NULL)',
        'INSERT INTO ' . TABLE_COLS . ' VALUES ("VariantOnTranscript/HGVS_Offset",                                          255, 100, 0, 1, 0, "HGVS Offset",                                   "", "HGVS Offset", "Indicates by how many bases the HGVS notations for this variant have been shifted.", "VARCHAR(200)", "HGVS Offset||text|6", "", "", 0, 0, 1, 0, NOW(), NULL, NULL)',
        'INSERT INTO ' . TABLE_COLS . ' VALUES ("VariantOnTranscript/Frequency/1000G/VEP/African",                          255, 100, 0, 1, 0, "1000G African Frequency (VEP)",                 "", "1000G African Frequency (VEP)", "Non-reference allele and frequency of existing variant in 1000 Genomes combined African population.", "VARCHAR(200)", "1000G African Frequency (VEP)||text|6", "", "", 0, 0, 1, 0, NOW(), NULL, NULL)',
        'INSERT INTO ' . TABLE_COLS . ' VALUES ("VariantOnTranscript/Frequency/1000G/VEP/American",                         255, 100, 0, 1, 0, "1000G American Frequency (VEP)",                "", "1000G American Frequency (VEP)", "Non-reference allele and frequency of existing variant in 1000 Genomes combined American population.", "VARCHAR(200)", "1000G American Frequency (VEP)||text|6", "", "", 0, 0, 1, 0, NOW(), NULL, NULL)',
        'INSERT INTO ' . TABLE_COLS . ' VALUES ("VariantOnTranscript/Frequency/1000G/VEP/East_Asian",                       255, 100, 0, 1, 0, "1000G East Asian Frequency (VEP)",              "", "1000G East Asian Frequency (VEP)", "Non-reference allele and frequency of existing variant in 1000 Genomes combined East Asian population.", "VARCHAR(200)", "1000G East Asian Frequency (VEP)||text|6", "", "", 0, 0, 1, 0, NOW(), NULL, NULL)',
        'INSERT INTO ' . TABLE_COLS . ' VALUES ("VariantOnTranscript/Frequency/1000G/VEP/European",                         255, 100, 0, 1, 0, "1000G European Frequency (VEP)",                "", "1000G European Frequency (VEP)", "Non-reference allele and frequency of existing variant in 1000 Genomes combined European population.", "VARCHAR(200)", "1000G European Frequency (VEP)||text|6", "", "", 0, 0, 1, 0, NOW(), NULL, NULL)',
        'INSERT INTO ' . TABLE_COLS . ' VALUES ("VariantOnTranscript/Frequency/1000G/VEP/South_Asian",                      255, 100, 0, 1, 0, "1000G South Asian Frequency (VEP)",             "", "1000G South Asian Frequency (VEP)", "Non-reference allele and frequency of existing variant in 1000 Genomes combined South Asian population.", "VARCHAR(200)", "1000G South Asian Frequency (VEP)||text|6", "", "", 0, 0, 1, 0, NOW(), NULL, NULL)',
        'INSERT INTO ' . TABLE_COLS . ' VALUES ("VariantOnTranscript/Frequency/1000G/VEP/African_American",                 255, 100, 0, 1, 0, "1000G Afr American Frequency (VEP)",            "", "1000G African American Frequency (VEP)", "Non-reference allele and frequency of existing variant in NHLBI-ESP African American population.", "VARCHAR(200)", "1000G African American Frequency (VEP)||text|6", "", "", 0, 0, 1, 0, NOW(), NULL, NULL)',
        'INSERT INTO ' . TABLE_COLS . ' VALUES ("VariantOnTranscript/Frequency/1000G/VEP/European_American",                255, 100, 0, 1, 0, "1000G Euro American Frequency (VEP)",           "", "1000G European American Frequency (VEP)", "Non-reference allele and frequency of existing variant in NHLBI-ESP European American population.", "VARCHAR(200)", "1000G European American Frequency (VEP)||text|6", "", "", 0, 0, 1, 0, NOW(), NULL, NULL)',
        'INSERT INTO ' . TABLE_COLS . ' VALUES ("VariantOnTranscript/Frequency/ExAC/Adjusted",                              255, 100, 0, 1, 0, "Adjusted ExAC Frequency",                       "", "Adjusted ExAC Frequency", "Adjusted frequency of existing variant in ExAC combined population.", "VARCHAR(200)", "Adjusted ExAC Frequency||text|6", "", "", 0, 0, 1, 0, NOW(), NULL, NULL)',
        'INSERT INTO ' . TABLE_COLS . ' VALUES ("VariantOnTranscript/Frequency/ExAC/African_American",                      255, 100, 0, 1, 0, "ExAC African American Frequency",               "", "ExAC African American Frequency", "Frequency of existing variant in ExAC African/American population.", "VARCHAR(200)", "ExAC African American Frequency||text|6", "", "", 0, 0, 1, 0, NOW(), NULL, NULL)',
        'INSERT INTO ' . TABLE_COLS . ' VALUES ("VariantOnTranscript/Frequency/ExAC/American",                              255, 100, 0, 1, 0, "ExAC American Frequency",                       "", "ExAC American Frequency", "Frequency of existing variant in ExAC American population.", "VARCHAR(200)", "ExAC American Frequency||text|6", "", "", 0, 0, 1, 0, NOW(), NULL, NULL)',
        'INSERT INTO ' . TABLE_COLS . ' VALUES ("VariantOnTranscript/Frequency/ExAC/East_Asian",                            255, 100, 0, 1, 0, "ExAC East Asian Frequency",                     "", "ExAC East Asian Frequency", "Frequency of existing variant in ExAC East Asian population.", "VARCHAR(200)", "ExAC East Asian Frequency||text|6", "", "", 0, 0, 1, 0, NOW(), NULL, NULL)',
        'INSERT INTO ' . TABLE_COLS . ' VALUES ("VariantOnTranscript/Frequency/ExAC/Finnish",                               255, 100, 0, 1, 0, "ExAC Finnish Frequency",                        "", "ExAC Finnish Frequency", "Frequency of existing variant in ExAC Finnish population.", "VARCHAR(200)", "ExAC Finnish Frequency||text|6", "", "", 0, 0, 1, 0, NOW(), NULL, NULL)',
        'INSERT INTO ' . TABLE_COLS . ' VALUES ("VariantOnTranscript/Frequency/ExAC/Non_Finnish",                           255, 100, 0, 1, 0, "ExAC Non-Finnish European Frequency",           "", "ExAC Non-Finnish European Frequency", "Frequency of existing variant in ExAC Non-Finnish European population.", "VARCHAR(200)", "ExAC Non-Finnish European Frequency||text|6", "", "", 0, 0, 1, 0, NOW(), NULL, NULL)',
        'INSERT INTO ' . TABLE_COLS . ' VALUES ("VariantOnTranscript/Frequency/ExAC/Other",                                 255, 100, 0, 1, 0, "ExAC Other Combined Frequency",                 "", "ExAC other combined populations frequency", "Frequency of existing variant in ExAC combined other combined populations.", "VARCHAR(200)", "ExAC other combined populations frequency||text|6", "", "", 0, 0, 1, 0, NOW(), NULL, NULL)',
        'INSERT INTO ' . TABLE_COLS . ' VALUES ("VariantOnTranscript/Frequency/ExAC/South_Asian",                           255, 100, 0, 1, 0, "ExAC South Asian Frequency",                    "", "ExAC South Asian Frequency", "Frequency of existing variant in ExAC South Asian population.", "VARCHAR(200)", "ExAC South Asian Frequency||text|6", "", "", 0, 0, 1, 0, NOW(), NULL, NULL)',
        'INSERT INTO ' . TABLE_COLS . ' VALUES ("VariantOnTranscript/Clinical_Significance",                                255, 100, 0, 1, 0, "Clinical Significance (dbNSFP)",                "", "Clinical Significance (dbNSFP)", "Clinical significance of variant from dbSNP.", "VARCHAR(200)", "Clinical Significance||text|6", "", "", 0, 0, 1, 0, NOW(), NULL, NULL)',
        'INSERT INTO ' . TABLE_COLS . ' VALUES ("VariantOnTranscript/Somatic_Status",                                       255, 100, 0, 1, 0, "Somatic Status ",                               "", "Somatic Status ", "Somatic status of existing variant(s).", "VARCHAR(200)", "Somatic Status ||text|6", "", "", 0, 0, 1, 0, NOW(), NULL, NULL)',
        'INSERT INTO ' . TABLE_COLS . ' VALUES ("VariantOnTranscript/Phenotype",                                            255, 100, 0, 1, 0, "Phenotype",                                     "", "Phenotype", "Indicates if existing variant is associated with a phenotype, disease or trait.", "VARCHAR(200)", "Phenotype||text|6", "", "", 0, 0, 1, 0, NOW(), NULL, NULL)',
        'INSERT INTO ' . TABLE_COLS . ' VALUES ("VariantOnTranscript/Pubmed",                                               255, 100, 0, 1, 0, "Pubmed",                                        "", "Pubmed", "Pubmed ID(s) of publications that cite existing variant.", "TEXT", "Pubmed||text|6", "", "", 0, 0, 1, 0, NOW(), NULL, NULL)',
        'INSERT INTO ' . TABLE_COLS . ' VALUES ("VariantOnTranscript/Prediction/Condel_Score",                              255, 100, 0, 1, 0, "Condel Score",                                  "", "Consensus Deleteriousness (Condel) score", "Consensus Deleteriousness (Condel) score.", "DECIMAL(3,2)", "Condel score||text|6", "", "", 0, 0, 1, 0, NOW(), NULL, NULL)',
        'INSERT INTO ' . TABLE_COLS . ' VALUES ("VariantOnTranscript/Frequency/1000G/dbNSFP/Allele_Count",                  255, 100, 0, 1, 0, "1000Gp1 Allele Count",                          "", "1000Gp1 Allele Count (dbNSFP)", "Alternative allele counts in the whole 1000 genomes phase 1 (1000Gp1) data..", "VARCHAR(200)", "1000Gp1 Allele Count (dnNSFP)||text|6", "", "", 0, 0, 1, 0, NOW(), NULL, NULL)',
        'INSERT INTO ' . TABLE_COLS . ' VALUES ("VariantOnTranscript/Frequency/1000G/dbNSFP",                               255, 100, 0, 1, 0, "1000Gp1 Frequency",                             "", "1000Gp1 Frequency (dbNSFP)", "long.", "VARCHAR(200)", "1000Gp1 Frequency (dbNSFP)||text|6", "", "", 0, 0, 1, 0, NOW(), NULL, NULL)',
        'INSERT INTO ' . TABLE_COLS . ' VALUES ("VariantOnTranscript/Frequency/1000G/dbNSFP/African/Allele_Count",          255, 100, 0, 1, 0, "1000Gp1 African Allele Count",                  "", "1000Gp1 African Allele Count", "Alternative allele counts in the 1000Gp1 African descendent samples..", "VARCHAR(200)", "1000Gp1 African Allele Count||text|6", "", "", 0, 0, 1, 0, NOW(), NULL, NULL)',
        'INSERT INTO ' . TABLE_COLS . ' VALUES ("VariantOnTranscript/Frequency/1000G/dbNSFP/African",                       255, 100, 0, 1, 0, "1000Gp1 African Frequency",                     "", "1000Gp1 African Frequency", "Alternative allele frequency in the 1000Gp1 African descendent samples..", "VARCHAR(200)", "1000Gp1 African Frequency||text|6", "", "", 0, 0, 1, 0, NOW(), NULL, NULL)',
        'INSERT INTO ' . TABLE_COLS . ' VALUES ("VariantOnTranscript/Frequency/1000G/dbNSFP/American/Allele_Count",         255, 100, 0, 1, 0, "1000Gp1 American Allele Count",                 "", "1000Gp1 American Allele Count", "Alternative allele counts in the 1000Gp1 American descendent samples..", "VARCHAR(200)", "1000Gp1 American Allele Count||text|6", "", "", 0, 0, 1, 0, NOW(), NULL, NULL)',
        'INSERT INTO ' . TABLE_COLS . ' VALUES ("VariantOnTranscript/Frequency/1000G/dbNSFP/American",                      255, 100, 0, 1, 0, "1000Gp1 American Frequency",                    "", "1000Gp1 American Frequency", "Alternative allele frequency in the 1000Gp1 American descendent samples..", "VARCHAR(200)", "1000Gp1 American Frequency||text|6", "", "", 0, 0, 1, 0, NOW(), NULL, NULL)',
        'INSERT INTO ' . TABLE_COLS . ' VALUES ("VariantOnTranscript/Frequency/1000G/dbNSFP/Asian/Allele_Count",            255, 100, 0, 1, 0, "1000Gp1 Asian Allele Count",                    "", "1000Gp1 Asian Allele Count", "Alternative allele counts in the 1000Gp1 Asian descendent samples..", "VARCHAR(200)", "1000Gp1 Asian Allele Count||text|6", "", "", 0, 0, 1, 0, NOW(), NULL, NULL)',
        'INSERT INTO ' . TABLE_COLS . ' VALUES ("VariantOnTranscript/Frequency/1000G/dbNSFP/Asian",                         255, 100, 0, 1, 0, "1000Gp1 Asian Frequency",                       "", "1000Gp1 Asian Frequency", "Alternative allele frequency in the 1000Gp1 Asian descendent samples..", "VARCHAR(200)", "1000Gp1 Asian Frequency||text|6", "", "", 0, 0, 1, 0, NOW(), NULL, NULL)',
        'INSERT INTO ' . TABLE_COLS . ' VALUES ("VariantOnTranscript/Frequency/1000G/dbNSFP/European/Allele_Count",         255, 100, 0, 1, 0, "1000Gp1 European Allele Count",                 "", "1000Gp1 European Allele Count", "Alternative allele counts in the 1000Gp1 European descendent samples..", "VARCHAR(200)", "1000Gp1 European Allele Count||text|6", "", "", 0, 0, 1, 0, NOW(), NULL, NULL)',
        'INSERT INTO ' . TABLE_COLS . ' VALUES ("VariantOnTranscript/Frequency/1000G/dbNSFP/European",                      255, 100, 0, 1, 0, "1000Gp1 European Frequency",                    "", "1000Gp1 European Frequency", "Alternative allele frequency in the 1000Gp1 European descendent samples..", "VARCHAR(200)", "1000Gp1 European Frequency||text|6", "", "", 0, 0, 1, 0, NOW(), NULL, NULL)',
        'INSERT INTO ' . TABLE_COLS . ' VALUES ("VariantOnTranscript/Prediction/CADD_Phredlike",                            255, 100, 0, 1, 0, "CADD Phred-like Score",                         "", "CADD Phred-like Score", "CADD phred-like score. This is phred-like rank score based on whole genome CADD raw scores. Please refer to Kircher et al. (2014) Nature Genetics 46(3):310-5 for details. The larger the score the more likely the SNP has damaging effect. Please note the following copyright statement for CADD: CADD scores (http://cadd.gs.washington.edu/) are Copyright 2013 University of Washington and Hudson-Alpha Institute for Biotechnology (all rights reserved) but are freely available for all academic, non-commercial applications. For commercial licensing information contact Jennifer McCullar (mccullaj@uw.edu).", "VARCHAR(200)", "CADD Phred-like Score||text|6", "", "", 0, 0, 1, 0, NOW(), NULL, NULL)',
        'INSERT INTO ' . TABLE_COLS . ' VALUES ("VariantOnTranscript/Prediction/CADD_Raw",                                  255, 100, 0, 1, 0, "CADD Raw Score",                                "", "CADD Raw Score", "CADD raw score for funtional prediction of a SNP. Please refer to Kircher et al. (2014) Nature Genetics 46(3):310-5 for details. The larger the score the more likely the SNP has damaging effect. Please note the following copyright statement for CADD: CADD scores (http://cadd.gs.washington.edu/) are Copyright 2013 University of Washington and Hudson-Alpha Institute for Biotechnology (all rights reserved) but are freely available for all academic, non-commercial applications. For commercial licensing information contact Jennifer McCullar (mccullaj@uw.edu)..", "VARCHAR(200)", "CADD Raw Score||text|6", "", "", 0, 0, 1, 0, NOW(), NULL, NULL)',
        'INSERT INTO ' . TABLE_COLS . ' VALUES ("VariantOnTranscript/Prediction/CADD_Raw_Ranked",                           255, 100, 0, 1, 0, "CADD Raw Score Ranked",                         "", "CADD Raw Score Ranked", "CADD raw scores were ranked among all CADD raw scores in dbNSFP. The rankscore is the ratio of the rank of the score over the total number of CADD raw scores in dbNSFP. Please note the following copyright statement for CADD: CADD scores (http://cadd.gs.washington.edu/) are Copyright 2013 University of Washington and Hudson-Alpha Institute for Biotechnology (all rights reserved) but are freely available for all academic, non-commercial applications. For commercial licensing information contact Jennifer McCullar (mccullaj@uw.edu)..", "VARCHAR(200)", "CADD Raw Score Ranked||text|6", "", "", 0, 0, 1, 0, NOW(), NULL, NULL)',
        'INSERT INTO ' . TABLE_COLS . ' VALUES ("VariantOnTranscript/Frequency/ESP6500/American",                           255, 100, 0, 1, 0, "ESP6500 African American Frequency",            "", "ESP6500 African American Frequency", "Alternative allele frequency in the Afrian American samples of the NHLBI GO Exome Sequencing Project (ESP6500 data set)..", "VARCHAR(200)", "ESP6500 African American Frequency||text|6", "", "", 0, 0, 1, 0, NOW(), NULL, NULL)',
        'INSERT INTO ' . TABLE_COLS . ' VALUES ("VariantOnTranscript/Frequency/ESP6500/European_American",                  255, 100, 0, 1, 0, "ESP6500 European American Frequency",           "", "ESP6500 European American Frequency", "Alternative allele frequency in the European American samples of the NHLBI GO Exome Sequencing Project (ESP6500 data set)..", "VARCHAR(200)", "ESP6500 European American Frequency||text|6", "", "", 0, 0, 1, 0, NOW(), NULL, NULL)',
        'INSERT INTO ' . TABLE_COLS . ' VALUES ("VariantOnTranscript/Prediction/FATHMM",                                    255, 100, 0, 1, 0, "FATHMM Prediction",                             "", "FATHMM Prediction", "If a FATHMMori score is <=-1.5 (or rankscore <=0.81415) the corresponding NS is predicted as D(AMAGING); otherwise it is predicted as T(OLERATED). Multiple predictions separated by ;.", "VARCHAR(200)", "FATHMM Prediction||text|6", "", "", 0, 0, 1, 0, NOW(), NULL, NULL)',
        'INSERT INTO ' . TABLE_COLS . ' VALUES ("VariantOnTranscript/Prediction/FATHMM_Ranked_Score",                       255, 100, 0, 1, 0, "FATHMM Ranked Score",                           "", "FATHMM Ranked Score", "FATHMMori scores were ranked among all FATHMMori scores in dbNSFP. The rankscore is the ratio of the rank of the score over the total number of FATHMMori scores in dbNSFP. If there are multiple scores, only the most damaging (largest) rankscore is presented. The scores range from 0 to 1.", "VARCHAR(200)", "FATHMM Ranked Score||text|6", "", "", 0, 0, 1, 0, NOW(), NULL, NULL)',
        'INSERT INTO ' . TABLE_COLS . ' VALUES ("VariantOnTranscript/Prediction/FATHMM_Score",                              255, 100, 0, 1, 0, "FATHMM Score",                                  "", "FATHMM Score", "FATHMM default score (weighted for human inherited-disease mutations with Disease Ontology) (FATHMMori). Scores range from -18.09 to 11.0. Multiple scores separated by ; Please refer to Shihab et al. (2013) Human Mutation 34(1):57-65 for details.", "VARCHAR(200)", "FATHMM Score||text|6", "", "", 0, 0, 1, 0, NOW(), NULL, NULL)',
        'INSERT INTO ' . TABLE_COLS . ' VALUES ("VariantOnTranscript/Prediction/GERP_Neutral_Rate",                         255, 100, 0, 1, 0, "GERP++ Neutral Rate",                           "", "GERP++ Neutral Rate", "GERP++ neutral rate.", "VARCHAR(200)", "GERP++ Neutral Rate||text|6", "", "", 0, 0, 1, 0, NOW(), NULL, NULL)',
        'INSERT INTO ' . TABLE_COLS . ' VALUES ("VariantOnTranscript/Prediction/GERP_Score",                                255, 100, 0, 1, 0, "GERP++ RS Score",                               "", "GERP++ RS Score", "GERP++ RS score, the larger the score, the more conserved the site..", "VARCHAR(200)", "GERP++ RS Score||text|6", "", "", 0, 0, 1, 0, NOW(), NULL, NULL)',
        'INSERT INTO ' . TABLE_COLS . ' VALUES ("VariantOnTranscript/Prediction/GERP_Ranked_Score",                         255, 100, 0, 1, 0, "GERP++ RS Ranked Score",                        "", "GERP++ RS Ranked Score", "GERP++ RS scores were ranked among all GERP++ RS scores in dbNSFP. The rankscore is the ratio of the rank of the score over the total number of GERP++ RS scores in dbNSFP..", "VARCHAR(200)", "GERP++ RS Ranked Score||text|6", "", "", 0, 0, 1, 0, NOW(), NULL, NULL)',
        'INSERT INTO ' . TABLE_COLS . ' VALUES ("VariantOnTranscript/Prediction/LRT_Omega",                                 255, 100, 0, 1, 0, "LRT Omega",                                     "", "LRT Omega", "estimated nonsynonymous-to-synonymous-rate ratio (Omega, reported by LRT).", "VARCHAR(200)", "LRT Omega||text|6", "", "", 0, 0, 1, 0, NOW(), NULL, NULL)',
        'INSERT INTO ' . TABLE_COLS . ' VALUES ("VariantOnTranscript/Prediction/LRT_Ranked_Score",                          255, 100, 0, 1, 0, "LRT Ranked Score",                              "", "LRT Ranked Score", "LRTori scores were first converted as LRTnew=1-LRTori*0.5 if Omega<1, or LRTnew=LRTori*0.5 if Omega>=1. Then LRTnew scores were ranked among all LRTnew scores in dbNSFP. The rankscore is the ratio of the rank over the total number of the scores in dbNSFP. The scores range from 0.00166 to 0.85682.", "VARCHAR(200)", "LRT Ranked Score||text|6", "", "", 0, 0, 1, 0, NOW(), NULL, NULL)',
        'INSERT INTO ' . TABLE_COLS . ' VALUES ("VariantOnTranscript/Prediction/LRT",                                       255, 100, 0, 1, 0, "LRT Prediction",                                "", "LRT Prediction", "LRT prediction, D(eleterious), N(eutral) or U(nknown), which is not solely determined by the score..", "VARCHAR(200)", "LRT Prediction||text|6", "", "", 0, 0, 1, 0, NOW(), NULL, NULL)',
        'INSERT INTO ' . TABLE_COLS . ' VALUES ("VariantOnTranscript/Prediction/LRT_Score",                                 255, 100, 0, 1, 0, "LRT Score",                                     "", "LRT Score", "The original LRT two-sided p-value (LRTori), ranges from 0 to 1..", "VARCHAR(200)", "LRT Score||text|6", "", "", 0, 0, 1, 0, NOW(), NULL, NULL)',
        'INSERT INTO ' . TABLE_COLS . ' VALUES ("VariantOnTranscript/Prediction/MetaLR",                                    255, 100, 0, 1, 0, "MetaLR Prediction",                             "", "MetaLR Prediction", "Prediction of our MetaLR based ensemble prediction score,T(olerated) or D(amaging). The score cutoff between D and T is 0.5. The rankscore cutoff between D and T is 0.82268..", "VARCHAR(200)", "MetaLR Prediction||text|6", "", "", 0, 0, 1, 0, NOW(), NULL, NULL)',
        'INSERT INTO ' . TABLE_COLS . ' VALUES ("VariantOnTranscript/Prediction/MetaLR_Ranked_Score",                       255, 100, 0, 1, 0, "MetaLR Ranked Score",                           "", "MetaLR Ranked Score", "MetaLR scores were ranked among all MetaLR scores in dbNSFP. The rankscore is the ratio of the rank of the score over the total number of MetaLR scores in dbNSFP. The scores range from 0 to 1..", "VARCHAR(200)", "MetaLR Ranked Score||text|6", "", "", 0, 0, 1, 0, NOW(), NULL, NULL)',
        'INSERT INTO ' . TABLE_COLS . ' VALUES ("VariantOnTranscript/Prediction/MetaLR_Score",                              255, 100, 0, 1, 0, "MetaLR Score",                                  "", "MetaLR Score", "Our logistic regression (LR) based ensemble prediction score, which incorporated 10 scores (SIFT, PolyPhen-2 HDIV, PolyPhen-2 HVAR, GERP++, MutationTaster, Mutation Assessor, FATHMM, LRT, SiPhy, PhyloP) and the maximum frequency observed in the 1000 genomes populations. Larger value means the SNV is more likely to be damaging. Scores range from 0 to 1.", "VARCHAR(200)", "MetaLR Score||text|6", "", "", 0, 0, 1, 0, NOW(), NULL, NULL)',
        'INSERT INTO ' . TABLE_COLS . ' VALUES ("VariantOnTranscript/Prediction/MetaSVM",                                   255, 100, 0, 1, 0, "MetaSVM Prediction",                            "", "MetaSVM Prediction", "Prediction of our SVM based ensemble prediction score,T(olerated) or D(amaging). The score cutoff between D and T is 0. The rankscore cutoff between D and T is 0.83357..", "VARCHAR(200)", "MetaSVM Prediction||text|6", "", "", 0, 0, 1, 0, NOW(), NULL, NULL)',
        'INSERT INTO ' . TABLE_COLS . ' VALUES ("VariantOnTranscript/Prediction/MetaSVM_Ranked_Score",                      255, 100, 0, 1, 0, "MetaSVM Ranked Score",                          "", "MetaSVM Ranked Score", "MetaSVM scores were ranked among all MetaSVM scores in dbNSFP. The rankscore is the ratio of the rank of the score over the total number of MetaSVM scores in dbNSFP. The scores range from 0 to 1..", "VARCHAR(200)", "MetaSVM Ranked Score||text|6", "", "", 0, 0, 1, 0, NOW(), NULL, NULL)',
        'INSERT INTO ' . TABLE_COLS . ' VALUES ("VariantOnTranscript/Prediction/MetaSVM_Score",                             255, 100, 0, 1, 0, "MetaSVM Score",                                 "", "MetaSVM Score", "Our support vector machine (SVM) based ensemble prediction score, which incorporated 10 scores (SIFT, PolyPhen-2 HDIV, PolyPhen-2 HVAR, GERP++, MutationTaster, Mutation Assessor, FATHMM, LRT, SiPhy, PhyloP) and the maximum frequency observed in the 1000 genomes populations. Larger value means the SNV is more likely to be damaging. Scores range from -2 to 3 in dbNSFP.", "VARCHAR(200)", "MetaSVM Score||text|6", "", "", 0, 0, 1, 0, NOW(), NULL, NULL)',
        'INSERT INTO ' . TABLE_COLS . ' VALUES ("VariantOnTranscript/Prediction/MutationAssessor",                          255, 100, 0, 1, 0, "Mutation Assessor Prediction",                  "", "Mutation Assessor Prediction", "MutationAssessor\'s functional impact of a variant : predicted functional, i.e. high (H) or medium (M), or predicted non-functional, i.e. low (L) or neutral (N). The MAori score cutoffs between H and M, M and L, and L and N, are 3.5, 1.935 and 0.8, respectively. The rankscore cutoffs between H and M, M and L, and L and N, are 0.92924, 0.51945 and 0.19692, respectively.", "VARCHAR(200)", "Mutation Assessor Prediction||text|6", "", "", 0, 0, 1, 0, NOW(), NULL, NULL)',
        'INSERT INTO ' . TABLE_COLS . ' VALUES ("VariantOnTranscript/Prediction/MutationAssessor_Ranked_Score",             255, 100, 0, 1, 0, "Mutation Assessor Ranked Score",                "", "Mutation Assessor Ranked Score", "MAori scores were ranked among all MAori scores in dbNSFP. The rankscore is the ratio of the rank of the score over the total number of MAori scores in dbNSFP. The scores range from 0 to 1..", "VARCHAR(200)", "Mutation Assessor Ranked Score||text|6", "", "", 0, 0, 1, 0, NOW(), NULL, NULL)',
        'INSERT INTO ' . TABLE_COLS . ' VALUES ("VariantOnTranscript/Prediction/MutationAssessor_Score",                    255, 100, 0, 1, 0, "Mutation Assessor Score",                       "", "Mutation Assessor Score", "MutationAssessor functional impact combined score (MAori). The score ranges from -5.135 to 6.49 in dbNSFP. Please refer to Reva et al. (2011) Nucl. Acids Res. 39(17):e118 for details..", "VARCHAR(200)", "Mutation Assessor Score||text|6", "", "", 0, 0, 1, 0, NOW(), NULL, NULL)',
        'INSERT INTO ' . TABLE_COLS . ' VALUES ("VariantOnTranscript/Prediction/MutationTaster_Ranked_Score",               255, 100, 0, 1, 0, "Mutation Taster Ranked Score",                  "", "Mutation Taster Ranked Score", "The MTori scores were first converted: if the prediction is A or D MTnew=MTori; if the prediction is N or P, MTnew=1-MTori. Then MTnew scores were ranked among all MTnew scores in dbNSFP. The rankscore is the ratio of the rank of the score over the total number of MTnew scores in dbNSFP. The scores range from 0.09067 to 0.80722..", "VARCHAR(200)", "Mutation Taster Ranked Score||text|6", "", "", 0, 0, 1, 0, NOW(), NULL, NULL)',
        'INSERT INTO ' . TABLE_COLS . ' VALUES ("VariantOnTranscript/Prediction/MutationTaster",                            255, 100, 0, 1, 0, "MutationTaster prediction",                     "", "MutationTaster prediction.", "MutationTaster prediction, A (disease_causing_automatic), D (disease_causing), N (polymorphism) or P (polymorphism_automatic). The score cutoff between D and N is 0.5 for MTori and 0.31655 for the rankscore.", "VARCHAR(200)", "MutationTaster prediction||select|1|true|false|false", "A\r\nD\r\nN\r\nP", "", 0, 0, 1, 0, NOW(), NULL, NULL)',
        'INSERT INTO ' . TABLE_COLS . ' VALUES ("VariantOnTranscript/Prediction/MutationTaster_Score",                      255, 100, 0, 1, 0, "MutationTaster score",                          "", "MutationTaster score.", "MutationTaster p-value (MTori), ranges from 0 to 1.", "VARCHAR(200)", "MutationTaster score||text|6", "", "", 0, 0, 1, 0, NOW(), NULL, NULL)',
        'INSERT INTO ' . TABLE_COLS . ' VALUES ("VariantOnTranscript/Prediction/PROVEAN_Ranked_Score",                      255, 100, 0, 1, 0, "PROVEAN Ranked Score",                          "", "PROVEAN Ranked Score", "PROVEANori were first converted to PROVEANnew=1-(PROVEANori+14)/28, then ranked among all PROVEANnew scores in dbNSFP. The rankscore is the ratio of the rank the PROVEANnew score over the total number of PROVEANnew scores in dbNSFP. If there are multiple scores, only the most damaging (largest) rankscore is presented..", "VARCHAR(200)", "PROVEAN Ranked Score||text|6", "", "", 0, 0, 1, 0, NOW(), NULL, NULL)',
        'INSERT INTO ' . TABLE_COLS . ' VALUES ("VariantOnTranscript/Prediction/PROVEAN",                                   255, 100, 0, 1, 0, "PROVEAN Prediction",                            "", "PROVEAN Prediction", "If PROVEANori <= -2.5 (rankscore>=0.59) the corresponding NS is predicted as D(amaging); otherwise it is predicted as N(eutral). Multiple predictions separated by ;.", "VARCHAR(200)", "PROVEAN Prediction||text|6", "", "", 0, 0, 1, 0, NOW(), NULL, NULL)',
        'INSERT INTO ' . TABLE_COLS . ' VALUES ("VariantOnTranscript/Prediction/PROVEAN_Score",                             255, 100, 0, 1, 0, "PROVEAN Score",                                 "", "PROVEAN Score", "PROVEAN score (PROVEANori). Scores range from -14 to 14. The smaller the score the more likely the SNP has damaging effect. Multiple scores separated by ;. Details can be found in DOI: 10.1371/journal.pone.0046688.", "VARCHAR(200)", "PROVEAN Score||text|6", "", "", 0, 0, 1, 0, NOW(), NULL, NULL)',
        'INSERT INTO ' . TABLE_COLS . ' VALUES ("VariantOnTranscript/Prediction/Polyphen2_HDIV",                            255, 100, 0, 1, 0, "Polyphen2 HDIV Prediction",                     "", "Polyphen2 HDIV Prediction", "long.", "VARCHAR(200)", "Polyphen2 HDIV Prediction||text|6", "", "", 0, 0, 1, 0, NOW(), NULL, NULL)',
        'INSERT INTO ' . TABLE_COLS . ' VALUES ("VariantOnTranscript/Prediction/Polyphen2_HDIV_Ranked_Score",               255, 100, 0, 1, 0, "Polyphen2 HDIV Ranked Score",                   "", "Polyphen2 HDIV Ranked Score", "Polyphen2 HDIV scores were first ranked among all HDIV scores in dbNSFP. The rankscore is the ratio of the rank the score over the total number of the scores in dbNSFP. If there are multiple scores, only the most damaging (largest) rankscore is presented. The scores range from 0.02656 to 0.89917..", "VARCHAR(200)", "Polyphen2 HDIV Ranked Score||text|6", "", "", 0, 0, 1, 0, NOW(), NULL, NULL)',
        'INSERT INTO ' . TABLE_COLS . ' VALUES ("VariantOnTranscript/Prediction/Polyphen2_HDIV_Score",                      255, 100, 0, 1, 0, "Polyphen2 HDIV Score",                          "", "Polyphen2 HDIV Score", "Polyphen2 score based on HumDiv, i.e. hdiv_prob. The score ranges from 0 to 1. Multiple entries separated by ;..", "VARCHAR(200)", "Polyphen2 HDIV Score||text|6", "", "", 0, 0, 1, 0, NOW(), NULL, NULL)',
        'INSERT INTO ' . TABLE_COLS . ' VALUES ("VariantOnTranscript/Prediction/Polyphen2_HVAR",                            255, 100, 0, 1, 0, "Polyphen2 HVAR Prediction",                     "", "Polyphen2 HVAR Prediction", "Polyphen2 prediction based on HumVar, D (porobably damaging, HVAR score in [0.909,1] or rankscore in [0.62955,0.9711]), P (possibly damaging, HVAR in [0.447,0.908] or rankscore in [0.44359,0.62885]) and B (benign, HVAR score in [0,0.446] or rankscore in [0.01281,0.44315]). Score cutoff for binary classification is 0.5 for HVAR score or 0.45998 for rankscore, i.e. the prediction is neutral if the HVAR score is smaller than 0.5 (rankscore is smaller than 0.45998), and deleterious if the HVAR score is larger than 0.5 (rankscore is larger than 0.45998). Multiple entries are separated by ;..", "VARCHAR(200)", "Polyphen2 HVAR Prediction||text|6", "", "", 0, 0, 1, 0, NOW(), NULL, NULL)',
        'INSERT INTO ' . TABLE_COLS . ' VALUES ("VariantOnTranscript/Prediction/Polyphen2_HVAR_Ranked_Score",               255, 100, 0, 1, 0, "Polyphen2 HVAR Ranked Score",                   "", "Polyphen2 HVAR Ranked Score", "Polyphen2 HVAR scores were first ranked among all HVAR scores in dbNSFP. The rankscore is the ratio of the rank the score over the total number of the scores in dbNSFP. If there are multiple scores, only the most damaging (largest) rankscore is presented. The scores range from 0.01281 to 0.9711..", "VARCHAR(200)", "Polyphen2 HVAR Ranked Score||text|6", "", "", 0, 0, 1, 0, NOW(), NULL, NULL)',
        'INSERT INTO ' . TABLE_COLS . ' VALUES ("VariantOnTranscript/Prediction/Polyphen2_HVAR_Score",                      255, 100, 0, 1, 0, "Polyphen2 HVAR Score",                          "", "Polyphen2 HVAR Score", "Polyphen2 score based on HumVar, i.e. hvar_prob. The score ranges from 0 to 1. Multiple entries separated by ;..", "VARCHAR(200)", "Polyphen2 HVAR Score||text|6", "", "", 0, 0, 1, 0, NOW(), NULL, NULL)',
        'INSERT INTO ' . TABLE_COLS . ' VALUES ("VariantOnTranscript/Prediction/MetaSVM_MetaLR_Reliability_Index",          255, 100, 0, 1, 0, "MetaSVM & MetaLR Reliability Index",            "", "MetaSVM & MetaLR Reliability Index", "Number of observed component scores (except the maximum frequency in the 1000 genomes populations) for MetaSVM and MetaLR. Ranges from 1 to 10. As MetaSVM and MetaLR scores are calculated based on imputed data, the less missing component scores, the higher the reliability of the scores and predictions..", "VARCHAR(200)", "MetaSVM & MetaLR Reliability Index||text|6", "", "", 0, 0, 1, 0, NOW(), NULL, NULL)',
        'INSERT INTO ' . TABLE_COLS . ' VALUES ("VariantOnTranscript/Prediction/SIFT_dbNSFP",                               255, 100, 0, 1, 0, "SIFT Prediction (dbNSFP)",                      "", "SIFT Prediction (dbNSFP)", "If SIFTori is smaller than 0.05 (rankscore>0.55) the corresponding NS is predicted as D(amaging); otherwise it is predicted as T(olerated). Multiple predictions separated by ;.", "VARCHAR(200)", "SIFT Prediction (dbNSFP)||text|6", "", "", 0, 0, 1, 0, NOW(), NULL, NULL)',
        'INSERT INTO ' . TABLE_COLS . ' VALUES ("VariantOnTranscript/Prediction/SiPhy29way_Score",                          255, 100, 0, 1, 0, "SiPhy 29way Score",                             "", "SiPhy 29way Score", "SiPhy score based on 29 mammals genomes. The larger the score, the more conserved the site..", "VARCHAR(200)", "SiPhy 29way Score||text|6", "", "", 0, 0, 1, 0, NOW(), NULL, NULL)',
        'INSERT INTO ' . TABLE_COLS . ' VALUES ("VariantOnTranscript/Prediction/SiPhy29way_Ranked_Score",                   255, 100, 0, 1, 0, "SiPhy 29way Ranked Score",                      "", "SiPhy 29way Ranked Score", "SiPhy_29way_logOdds scores were ranked among all SiPhy_29way_logOdds scores in dbNSFP. The rankscore is the ratio of the rank of the score over the total number of SiPhy_29way_logOdds scores in dbNSFP..", "VARCHAR(200)", "SiPhy 29way Ranked Score||text|6", "", "", 0, 0, 1, 0, NOW(), NULL, NULL)',
        'INSERT INTO ' . TABLE_COLS . ' VALUES ("VariantOnTranscript/Prediction/SiPhy29way_Distribution",                   255, 100, 0, 1, 0, "SiPhy 29way Distribution",                      "", "SiPhy 29way Distribution", "The estimated stationary distribution of A, C, G and T at the site, using SiPhy algorithm based on 29 mammals genomes..", "VARCHAR(200)", "SiPhy 29way Distribution||text|6", "", "", 0, 0, 1, 0, NOW(), NULL, NULL)',
        'INSERT INTO ' . TABLE_COLS . ' VALUES ("VariantOnTranscript/UniSNP_IDs",                                           255, 100, 0, 1, 0, "UniSNP IDs",                                    "", "UniSNP IDs", "rs numbers from UniSNP, which is a cleaned version of dbSNP build 129, in format: rs number1;rs number2;....", "VARCHAR(200)", "UniSNP IDs||text|6", "", "", 0, 0, 1, 0, NOW(), NULL, NULL)',
        'INSERT INTO ' . TABLE_COLS . ' VALUES ("VariantOnTranscript/Prediction/VEST3_Ranked_Score",                        255, 100, 0, 1, 0, "VEST3 Ranked Score",                            "", "VEST3 Ranked Score", "VEST3 scores were ranked among all VEST3 scores in dbNSFP. The rankscore is the ratio of the rank of the score over the total number of VEST3 scores in dbNSFP. The scores range from 0 to 1. Please note VEST score is free for non-commercial use. For more details please refer tohttp://wiki.chasmsoftware.org/index.php/SoftwareLicense. Commercial users should contact the Johns Hopkins Technology Transfer office..", "VARCHAR(200)", "VEST3 Ranked Score||text|6", "", "", 0, 0, 1, 0, NOW(), NULL, NULL)',
        'INSERT INTO ' . TABLE_COLS . ' VALUES ("VariantOnTranscript/Prediction/VEST3_Score",                               255, 100, 0, 1, 0, "VEST3 Score",                                   "", "VEST3 Score", "VEST 3.0 score. Score ranges from 0 to 1. The larger the score the more likely the mutation may cause functional change. In case there are multiple scores for the same variant, the largest score (most damaging) is presented. Please refer to Carter et al., (2013) BMC Genomics. 14(3) 1-16 for details. Please note this score is free for non-commercial use. For more details please refer to http://wiki.chasmsoftware.org/index.php/SoftwareLicense. Commercial users should contact the Johns Hopkins Technology Transfer office..", "VARCHAR(200)", "VEST3 Score||text|6", "", "", 0, 0, 1, 0, NOW(), NULL, NULL)',
        'INSERT INTO ' . TABLE_COLS . ' VALUES ("VariantOnTranscript/Prediction/phastCons100way_Vert_Score",                255, 100, 0, 1, 0, "phastCons 100way Vertebrate Score",             "", "phastCons 100way Vertebrate Score", ".", "VARCHAR(200)", "phastCons 100way Vertebrate Score||text|6", "", "", 0, 0, 1, 0, NOW(), NULL, NULL)',
        'INSERT INTO ' . TABLE_COLS . ' VALUES ("VariantOnTranscript/Prediction/phastCons100way_Vert_Ranked_Score",         255, 100, 0, 1, 0, "phastCons 100way Vertebrate Ranked Score",      "", "phastCons 100way Vertebrate Ranked Score", "phastCons100way_vertebrate scores were ranked among all phastCons100way_vertebrate scores in dbNSFP. The rankscore is the ratio of the rank of the score over the total number of phastCons100way_vertebrate scores in dbNSFP..", "VARCHAR(200)", "phastCons 100way Vertebrate Ranked Score||text|6", "", "", 0, 0, 1, 0, NOW(), NULL, NULL)',
        'INSERT INTO ' . TABLE_COLS . ' VALUES ("VariantOnTranscript/Prediction/phastCons46way_Plac_Score",                 255, 100, 0, 1, 0, "phastCons 100way Placental Score",              "", "phastCons 100way Placental Score", "phastCons conservation score based on the multiple alignments of 33 placental mammal genomes (including human). The larger the score, the more conserved the site..", "VARCHAR(200)", "phastCons 100way Placental Score||text|6", "", "", 0, 0, 1, 0, NOW(), NULL, NULL)',
        'INSERT INTO ' . TABLE_COLS . ' VALUES ("VariantOnTranscript/Prediction/phastCons46way_Plac_Ranked_Score",          255, 100, 0, 1, 0, "phastCons 100way Placental Ranked Score",       "", "phastCons 100way Placental Ranked Score", "phastCons46way_placental scores were ranked among all phastCons46way_placental scores in dbNSFP. The rankscore is the ratio of the rank of the score over the total number of phastCons46way_placental scores in dbNSFP..", "VARCHAR(200)", "phastCons 100way Placental Ranked Score||text|6", "", "", 0, 0, 1, 0, NOW(), NULL, NULL)',
        'INSERT INTO ' . TABLE_COLS . ' VALUES ("VariantOnTranscript/Prediction/phastCons46way_Prim_Score",                 255, 100, 0, 1, 0, "phastCons 100way Primate Score",                "", "phastCons 100way Primate Score", "phastCons conservation score based on the multiple alignments of 10 primate genomes (including human). The larger the score, the more conserved the site..", "VARCHAR(200)", "phastCons 100way Primate Score||text|6", "", "", 0, 0, 1, 0, NOW(), NULL, NULL)',
        'INSERT INTO ' . TABLE_COLS . ' VALUES ("VariantOnTranscript/Prediction/phastCons46way_Prim_Ranked_Score",          255, 100, 0, 1, 0, "phastCons 100way Primate Ranked Score",         "", "phastCons 100way Primate Ranked Score", "phastCons46way_primate scores were ranked among all phastCons46way_primate scores in dbNSFP. The rankscore is the ratio of the rank of the score over the total number of phastCons46way_primate scores in dbNSFP..", "VARCHAR(200)", "phastCons 100way Primate Ranked Score||text|6", "", "", 0, 0, 1, 0, NOW(), NULL, NULL)',
        'INSERT INTO ' . TABLE_COLS . ' VALUES ("VariantOnTranscript/Prediction/phyloP100way_Vert_Score",                   255, 100, 0, 1, 0, "phyloP46way Vertebrate Score",                  "", "phyloP46way Vertebrate Score", "phyloP (phylogenetic p-values) conservation score based on the multiple alignments of 100 vertebrate genomes (including human). The larger the score, the more conserved the site..", "VARCHAR(200)", "phyloP46way Vertebrate Score||text|6", "", "", 0, 0, 1, 0, NOW(), NULL, NULL)',
        'INSERT INTO ' . TABLE_COLS . ' VALUES ("VariantOnTranscript/Prediction/phyloP100way_Vert_Ranked_Score",            255, 100, 0, 1, 0, "phyloP46way Vertebrate Ranked Score",           "", "phyloP46way Vertebrate Ranked Score", "phyloP100way_vertebrate scores were ranked among all phyloP100way_vertebrate scores in dbNSFP. The rankscore is the ratio of the rank of the score over the total number of phyloP100way_vertebrate scores in dbNSFP..", "VARCHAR(200)", "phyloP46way Vertebrate Ranked Score||text|6", "", "", 0, 0, 1, 0, NOW(), NULL, NULL)',
        'INSERT INTO ' . TABLE_COLS . ' VALUES ("VariantOnTranscript/Prediction/phyloP46way_Plac_Score",                    255, 100, 0, 1, 0, "phyloP46way Placental Score",                   "", "phyloP46way Placental Score", "phyloP (phylogenetic p-values) conservation score based on the multiple alignments of 33 placental mammal genomes (including human). The larger the score, the more conserved the site..", "VARCHAR(200)", "phyloP46way Placental Score||text|6", "", "", 0, 0, 1, 0, NOW(), NULL, NULL)',
        'INSERT INTO ' . TABLE_COLS . ' VALUES ("VariantOnTranscript/Prediction/phyloP46way_Plac_Ranked_Score",             255, 100, 0, 1, 0, "phyloP46way Placental Ranked Score",            "", "phyloP46way Placental Ranked Score", "phyloP46way_placental scores were ranked among all phyloP46way_placental scores in dbNSFP. The rankscore is the ratio of the rank of the score over the total number of phyloP46way_placental scores in dbNSFP..", "VARCHAR(200)", "phyloP46way Placental Ranked Score||text|6", "", "", 0, 0, 1, 0, NOW(), NULL, NULL)',
        'INSERT INTO ' . TABLE_COLS . ' VALUES ("VariantOnTranscript/Prediction/phyloP46way_Prim_Score",                    255, 100, 0, 1, 0, "phyloP46way Primate Score",                     "", "phyloP46way Primate Score", "phyloP (phylogenetic p-values) conservation score based on the multiple alignments of 10 primate genomes (including human). The larger the score, the more conserved the site..", "VARCHAR(200)", "phyloP46way Primate Score||text|6", "", "", 0, 0, 1, 0, NOW(), NULL, NULL)',
        'INSERT INTO ' . TABLE_COLS . ' VALUES ("VariantOnTranscript/Prediction/phyloP46way_Prim_Ranked_Score",             255, 100, 0, 1, 0, "phyloP46way Primate Ranked Score",              "", "phyloP46way Primate Ranked Score", "phyloP46way_primate scores were ranked among all phyloP46way_primate scores in dbNSFP. The rankscore is the ratio of the rank of the score over the total number of phyloP46way_primate scores in dbNSFP..", "VARCHAR(200)", "phyloP46way Primate Ranked Score||text|6", "", "", 0, 0, 1, 0, NOW(), NULL, NULL)',
        'INSERT INTO ' . TABLE_COLS . ' VALUES ("VariantOnTranscript/Prediction/Grantham",                                  255, 100, 0, 1, 0, "Grantham Score",                                "", "Grantham Score", "Grantham Matrix score - Grantham, R. Amino Acid Difference Formula to Help Explain Protein Evolution, Science 1974 Sep 6;185(4154):862-4..", "VARCHAR(200)", "Grantham Score||text|6", "", "", 0, 0, 1, 0, NOW(), NULL, NULL)',


        //'INSERT INTO ' . TABLE_COLS . ' VALUES ("LOVD",            255, 100, 0, 1, 0, "short",           "", "short", "long.", "VARCHAR(255)", "short||text|6", "", "", 0, 0, 1, 0, NOW(), NULL, NULL)',

    );
    // Add these new columns to the existing columns.
    //$aColSQL = array_merge($aColSQL, $aMGHAColSQL);
}else{

    $aColSQL =
        array(
            //                                       COL ID,                                   ORDER,WIDTH,HGVS,STND,MAND.,TITLE,             FORM DESCRIPTION,                                                                                                               SHORT LEGEND,                                                                                        FULL LEGEND,                                                                                                                                                                                                                                                                                       MYSQL          FORM DEFINITION                                                                                                                                                                                                         SELECT OPTIONS, REGEXP,                                 PUBLIC (VIEW,ADD), COUNT?, CREATED_BY, CREATED DATE, EDITED BY, EDITED DATE.
            'INSERT INTO ' . TABLE_COLS . ' VALUES ("Individual/Age_of_death",                    248, 100, 0, 0, 0, "Age of death",         "Type 35y for 35 years, 04y08m for 4 years and 8 months, 18y? for around 18 years, >54y for still alive at 55, ? for unknown.", "The age at which the individual deceased, if known and applicable. 04y08m = 4 years and 8 months.", "The age at which the individual deceased, if known and applicable.\r\n<UL style=\"margin-top:0px;\">\r\n  <LI>35y = 35 years</LI>\r\n  <LI>04y08m = 4 years and 8 months</LI>\r\n  <LI>18y? = around 18 years</LI>\r\n  <LI>&gt;54y = still alive at 55</LI>\r\n  <LI>? = unknown</LI>\r\n</UL>", "VARCHAR(12)", "Age of death|The age at which the individual deceased, if known and applicable. Numbers lower than 10 should be prefixed by a zero and the field should always begin with years, to facilitate sorting on this column.|text|10", "", "/^([<>]?\\\\d{2,3}y(\\\\d{2}m(\\\\d{2}d)?)?)?\\\\??$/", 1, 1, 1, 0, NOW(), NULL, NULL)',
            'INSERT INTO ' . TABLE_COLS . ' VALUES ("Individual/Consanguinity",                   240,  40, 0, 0, 0, "Consanguinity",        "Indicates whether the parents are related (consanguineous), not related (non-consanguineous) or whether consanguinity is not known (unknown)", "Indicates whether the parents are related (consanguineous), not related (non-consanguineous) or whether consanguinity is not known (unknown)", "Indicates whether the parents are related (consanguineous), not related (non-consanguineous) or whether consanguinity is not known (unknown)", "VARCHAR(100)", "Consanguinity||select|1|--Not specified--|false|false", "? = Unknown\r\nno = Non-consanguineous parents\r\nyes = Consanguineous parents", "", 1, 1, 1, 0, NOW(), NULL, NULL)',
            'INSERT INTO ' . TABLE_COLS . ' VALUES ("Individual/Death/Cause",                     249, 150, 0, 0, 0, "Cause of death",       "", "The cause of the individual\'s death, if known and applicable.", "The cause of the individual\'s death, if known and applicable.", "VARCHAR(255)", "Cause of death|The cause of the individual\'s death, if known and applicable.|text|30", "", "", 1, 1, 1, 0, NOW(), NULL, NULL)',
            // SEER list: 1 Male, 2 Female, 3 Other (hermaphrodite, sex chromosome aberrations), 4 Transsexual/Transgender, 9 Not stated/Unknown
            'INSERT INTO ' . TABLE_COLS . ' VALUES ("Individual/Gender",                            3,  70, 0, 1, 0, "Gender",               "", "Individual\'s gender.", "The gender of the reported individual.", "VARCHAR(100)", "Gender||select|1|--Not specified--|false|false", "? = Unknown\r\nF = Female\r\nM = Male\r\nrF = Raised as female\r\nrM = Raised as male", "", 1, 1, 1, 0, NOW(), NULL, NULL)',
            'INSERT INTO ' . TABLE_COLS . ' VALUES ("Individual/Lab_ID",                            1,  80, 1, 1, 1, "Lab-ID",               "", "The individual\'s ID in the hospital, diagnostic laboratory or in the publication.", "The individual\'s ID in the hospital, diagnostic laboratory or in the publication.", "VARCHAR(20)", "Lab ID||text|20", "", "", 0, 1, 0, 0, NOW(), NULL, NULL)',
            'INSERT INTO ' . TABLE_COLS . ' VALUES ("Individual/Origin/Ethnic",                   201, 200, 0, 0, 0, "Ethnic origin",        "Ethnic origin of individual; e.g. African, Caucasian, gypsy, jew (Ashkenazi).", "Ethnic origin of individual; e.g. African, Caucasian, gypsy, jew (Ashkenazi).", "The ethnic origin of the individual; e.g. African, Caucasian, gypsy, jew (Ashkenazi).", "VARCHAR(50)", "Ethnic origin|If mixed, please indicate origin of father and mother, if known.|text|20", "", "", 1, 1, 1, 0, NOW(), NULL, NULL)',
            'INSERT INTO ' . TABLE_COLS . ' VALUES ("Individual/Origin/Geographic",               200, 200, 0, 0, 0, "Geographic origin",    "Geographic origin of individual; Belgium = individual\'s origin is Belgium, (France) = reported by laboratory in France, individual\'s origin not sure.", "Geographic origin of individual; Belgium = individual\'s origin is Belgium, (France) = reported by laboratory in France, individual\'s origin not sure.", "The geographic origin of the individual (country and/or region); Belgium = individual\'s origin is Belgium, (France) = reported by laboratory in France, individual\'s origin not sure.", "VARCHAR(50)", "Geographic origin|If mixed, please indicate origin of father and mother, if known.|text|30", "", "", 1, 1, 1, 0, NOW(), NULL, NULL)',
            'INSERT INTO ' . TABLE_COLS . ' VALUES ("Individual/Origin/Population",               202, 200, 0, 0, 0, "Population",           "", "Individual population.", "Additional information on the individual\'s population.", "VARCHAR(50)", "Individual population||text|30", "", "", 1, 1, 1, 0, NOW(), NULL, NULL)',
            'INSERT INTO ' . TABLE_COLS . ' VALUES ("Individual/Reference",                         2, 200, 0, 0, 0, "Reference",            "", "Reference to publication describing the individual/family.", "Reference to publication describing the individual/family, possibly giving more phenotypic details than listed in this database entry, including link to PubMed or other source, e.g. \"den Dunnen ASHG2003 P2346\". References in the &quot;Country:City&quot; format indicate that the variant was submitted directly to this database by the laboratory indicated.", "VARCHAR(200)", "Reference||text|50", "", "", 1, 1, 1, 0, NOW(), NULL, NULL)',
            'INSERT INTO ' . TABLE_COLS . ' VALUES ("Individual/Remarks",                         250, 200, 0, 1, 0, "Remarks",              "", "Remarks about the individual.", "Remarks about the individual.", "TEXT", "Remarks|Only provide data that do not belong in any other field.|textarea|50|3", "", "", 1, 1, 1, 0, NOW(), NULL, NULL)',
            'INSERT INTO ' . TABLE_COLS . ' VALUES ("Individual/Remarks_Non_Public",              251, 200, 0, 1, 0, "Remarks (non public)", "", "Non-public remarks about the individual.", "Non-public remarks about the individual.", "TEXT", "Remarks (non public)||textarea|50|3", "", "", 0, 0, 0, 0, NOW(), NULL, NULL)',
            'INSERT INTO ' . TABLE_COLS . ' VALUES ("Phenotype/Additional",                       250, 200, 0, 1, 0, "Phenotype details",    "Additional information on the phenotype of the individual.", "Additional information on the phenotype of the individual.", "Additional information on the phenotype of the individual.", "TEXT", "Additional phenotype details||textarea|40|4", "", "", 1, 1, 1, 0, NOW(), NULL, NULL)',
            // If you ever change this column, change the install script accordingly.
            'INSERT INTO ' . TABLE_COLS . ' VALUES ("Phenotype/Age",                               10, 100, 0, 0, 0, "Age examined",         "Type 35y for 35 years, 04y08m for 4 years and 8 months, 18y? for around 18 years, >54y for older than 54, ? for unknown.", "The age at which the individual was examined, if known. 04y08m = 4 years and 8 months.", "The age at which the individual was examined, if known.\r\n<UL style=\"margin-top:0px;\">\r\n  <LI>35y = 35 years</LI>\r\n  <LI>04y08m = 4 years and 8 months</LI>\r\n  <LI>18y? = around 18 years</LI>\r\n  <LI>&gt;54y = older than 54</LI>\r\n  <LI>? = unknown</LI>\r\n</UL>", "VARCHAR(12)", "Age at examination|The age at which the individual was examined, if known. Numbers lower than 10 should be prefixed by a zero and the field should always begin with years, to facilitate sorting on this column.|text|10", "", "/^([<>]?\\\\d{2,3}y(\\\\d{2}m(\\\\d{2}d)?)?)?\\\\??$/", 1, 1, 1, 0, NOW(), NULL, NULL)',
            'INSERT INTO ' . TABLE_COLS . ' VALUES ("Phenotype/Age/Onset",                          1, 100, 0, 0, 0, "Age of onset",         "Type 35y for 35 years, 04y08m for 4 years and 8 months, 18y? for around 18 years, >54y for older than 54, ? for unknown.", "The age at which the first symptoms of the disease appeared in the individual, if known. 04y08m = 4 years and 8 months.", "The age at which the first symptoms appeared in the individual, if known.\r\n<UL style=\"margin-top:0px;\">\r\n  <LI>35y = 35 years</LI>\r\n  <LI>04y08m = 4 years and 8 months</LI>\r\n  <LI>18y? = around 18 years</LI>\r\n  <LI>&gt;54y = older than 54</LI>\r\n  <LI>? = unknown</LI>\r\n</UL>", "VARCHAR(12)", "Age of onset|The age at which the first symptoms appeared in the individual, if known. Numbers lower than 10 should be prefixed by a zero and the field should always begin with years, to facilitate sorting on this column.|text|10", "", "/^([<>]?\\\\d{2,3}y(\\\\d{2}m(\\\\d{2}d)?)?)?\\\\??$/", 1, 1, 1, 0, NOW(), NULL, NULL)',
            'INSERT INTO ' . TABLE_COLS . ' VALUES ("Phenotype/Date",                               5,  80, 0, 0, 0, "Date",                 "Format: YYYY-MM-DD.", "Date the phenotype was observed.", "Date the phenotype was observed, in YYYY-MM-DD format.", "DATE", "Date||text|10", "", "", 1, 1, 1, 0, NOW(), NULL, NULL)',
            'INSERT INTO ' . TABLE_COLS . ' VALUES ("Phenotype/Inheritance",                      254, 200, 1, 1, 0, "Inheritance",          "Indicates the inheritance of the phenotype in the family; unknown, familial (autosomal/X-linked, dominant/ recessive), paternal (Y-linked), maternal (mitochondrial), isolated (sporadic) or complex", "Indicates the inheritance of the phenotype in the family; unknown, familial (autosomal/X-linked, dominant/ recessive), paternal (Y-linked), maternal (mitochondrial), isolated (sporadic) or complex", "Indicates the inheritance of the phenotype in the family; unknown, familial (autosomal/X-linked, dominant/ recessive), paternal (Y-linked), maternal (mitochondrial), isolated (sporadic) or complex", "VARCHAR(50)", "Inheritance||select|1|--Not specified--|false|false", "Unknown\r\nFamilial\r\nFamilial, autosomal dominant\r\nFamilial, autosomal recessive\r\nFamilial, X-linked dominant\r\nFamilial, X-linked dominant, male sparing\r\nFamilial, X-linked recessive\r\nPaternal, Y-linked\r\nMaternal, mitochondrial\r\nIsolated (sporadic)\r\nComplex", "", 1, 1, 1, 0, NOW(), NULL, NULL)',
            // If you ever change this column, change the install script accordingly.
            'INSERT INTO ' . TABLE_COLS . ' VALUES ("Phenotype/Length",                           200, 100, 0, 0, 0, "Length",               "", "Length of the individual, in cm.", "Length of the individual, in centimeters (cm).", "SMALLINT(3) UNSIGNED", "Length of individual (cm)|Length of individual, in centimeters.|text|3", "", "", 1, 1, 1, 0, NOW(), NULL, NULL)',
            'INSERT INTO ' . TABLE_COLS . ' VALUES ("Screening/Date",                               1,  80, 0, 0, 0, "Date",                 "Format: YYYY-MM-DD.", "Date the detection technique was performed.", "Date the detection technique was performed, in YYYY-MM-DD format.", "DATE", "Date||text|10", "", "", 1, 1, 1, 0, NOW(), NULL, NULL)',
            'INSERT INTO ' . TABLE_COLS . ' VALUES ("Screening/Technique",                          3, 200, 1, 1, 1, "Technique",            "Technique(s) used to identify the sequence variant; select multiple when more were used. For SEQ-NG, please indicate in the variant\'s remarks field the number of reads showing the variant (e.g. 47/96 reads, 123/123 reads), and create a second screening if you confirmed a variant using another method (e.g. SEQ).", "Technique(s) used to identify the sequence variant.", "Technique(s) used to identify the sequence variant.", "TEXT", "Technique(s) used||select|5|false|true|false", "? = Unknown\r\narrayCGH = array for Comparative Genomic Hybridisation\r\narraySEQ = array for resequencing\r\narraySNP = array for SNP typing\r\narrayCNV = array for Copy Number Variation (SNP and CNV probes)\r\nBESS = Base Excision Sequence Scanning\r\nCMC = Chemical Mismatch Cleavage\r\nCSCE = Conformation Sensitive Capillary Electrophoresis\r\nDGGE = Denaturing-Gradient Gel-Electrophoresis\r\nDHPLC = Denaturing High-Performance Liquid Chromatography\r\nDOVAM = Detection Of Virtually All Mutations (SSCA variant)\r\nddF = dideoxy Fingerprinting\r\nDSCA = Double-Strand DNA Conformation Analysis\r\nEMC = Enzymatic Mismatch Cleavage\r\nHD = HeteroDuplex analysis\r\nMCA = high-resolution Melting Curve Analysis (hrMCA)\r\nIHC = Immuno-Histo-Chemistry\r\nMAPH = Multiplex Amplifiable Probe Hybridisation\r\nMLPA = Multiplex Ligation-dependent Probe Amplification\r\nSEQ-NG = Next-Generation Sequencing\r\nSEQ-NG-H = Next-Generation Sequencing - Helicos\r\nSEQ-NG-I = Next-Generation Sequencing - Illumina/Solexa\r\nSEQ-NG-R = Next-Generation Sequencing - Roche/454\r\nSEQ-NG-S = Next-Generation Sequencing - SOLiD\r\nNorthern = Northern blotting\r\nPCR = Polymerase Chain Reaction\r\nPCRdig = PCR + restriction enzyme digestion\r\nPCRlr = PCR, long-range\r\nPCRm = PCR, multiplex\r\nPCRq = PCR, quantitative\r\nPAGE = Poly-Acrylamide Gel-Electrophoresis\r\nPTT = Protein Truncation Test\r\nPFGE = Pulsed-Field Gel-Electrophoresis (+Southern)\r\nRT-PCR = Reverse Transcription and PCR\r\nSEQ = SEQuencing\r\nSBE = Single Base Extension\r\nSSCA = Single-Strand DNA Conformation polymorphism Analysis (SSCP)\r\nSSCAf = SSCA, fluorescent (SSCP)\r\nSouthern = Southern blotting\r\nTaqMan = TaqMan assay\r\nWestern = Western Blotting", "", 1, 1, 1, 0, NOW(), NULL, NULL)',
            'INSERT INTO ' . TABLE_COLS . ' VALUES ("Screening/Template",                           2,  80, 1, 1, 1, "Template",             "", "Template(s) used to detect the sequence variant; DNA = genomic DNA, RNA = RNA (cDNA).", "Template(s) used to detect the sequence variant; DNA = genomic DNA, RNA = RNA (cDNA).", "TEXT", "Detection template||select|3|false|true|false", "DNA\r\nRNA = RNA (cDNA)\r\nProtein\r\n? = unknown", "", 1, 1, 1, 0, NOW(), NULL, NULL)',
            'INSERT INTO ' . TABLE_COLS . ' VALUES ("Screening/Tissue",                             4, 100, 0, 0, 1, "Tissue",               "", "Tissue type used for the detection of sequence variants.", "Tissue type used for the detection of sequence variants.", "VARCHAR(25)", "Tissue||text|20", "", "", 1, 1, 1, 0, NOW(), NULL, NULL)',
            'INSERT INTO ' . TABLE_COLS . ' VALUES ("VariantOnGenome/Conservation_score/GERP",      4, 100, 0, 0, 0, "GERP conservation",    "", "Conservation score as calculated by GERP.", "The Conservation score as calculated by GERP.", "DECIMAL(5,3)", "GERP conservation score||text|6", "", "", 1, 1, 1, 0, NOW(), NULL, NULL)',
            'INSERT INTO ' . TABLE_COLS . ' VALUES ("VariantOnGenome/DBID",                         7, 120, 1, 1, 1, "DB-ID",                "NOTE: This field will be predicted and filled in by LOVD, if left empty.", "Database ID of variant starting with the HGNC gene symbol, followed by an underscore (_) and a six digit number (e.g. DMD_012345). _000000 is used for variants where DNA was not analysed (change predicted from RNA analysis), variants seen in animal models or variants not seen in humans but functionally tested in vitro.", "Database ID of variant, grouping multiple observations of the same variant together, starting with the HGNC gene symbol, followed by an underscore (_) and a six digit number (e.g. DMD_012345). _000000 is used for variants where DNA was not analysed (change predicted from RNA analysis), variants seen in animal models or variants not seen in humans but functionally tested in vitro.", "VARCHAR(50)", "ID|This ID is used to group multiple observations of the same variant together. This field will be predicted and filled in by LOVD when left empty. The ID starts with the HGNC gene symbol of the transcript most influenced by the variant or otherwise the closest gene, followed by an underscore (_) and a six digit number (e.g. DMD_012345). _000000 is used for variants where DNA was not analysed (change predicted from RNA analysis), variants seen in animal models or variants not seen in humans but functionally tested in vitro.|text|20", "", "/^(chr(\\\\d{1,2}|[XYM])|(C(\\\\d{1,2}|[XYM])orf[\\\\d][\\\\dA-Z]*-|[A-Z][A-Z0-9]+-)?(C(\\\\d{1,2}|[XYM])orf[\\\\d][\\\\dA-Z]*|[A-Z][A-Z0-9-]+))_\\\\d{6}$/", 1, 0, 1, 0, NOW(), NULL, NULL)',
            'INSERT INTO ' . TABLE_COLS . ' VALUES ("VariantOnGenome/dbSNP",                        8, 120, 0, 1, 0, "dbSNP ID",             "", "The dbSNP ID.", "The dbSNP ID.", "VARCHAR(15)", "dbSNP ID|If available, please fill in the dbSNP ID, such as rs12345678.|text|10", "", "/^[rs]s\\\\d+$/", 0, 0, 1, 0, NOW(), NULL, NULL)',
            'INSERT INTO ' . TABLE_COLS . ' VALUES ("VariantOnGenome/DNA",                          2, 200, 1, 1, 1, "DNA change (genomic)", "", "Description of variant at DNA level, based on the genomic DNA reference sequence (following HGVS recommendations).", "Description of variant at DNA level, based on the genomic DNA reference sequence (following HGVS recommendations).<BR>\r\n<UL style=\"margin-top : 0px;\">\r\n  <LI>g.12345678C>T</LI>\r\n  <LI>g.12345678_12345890del</LI>\r\n  <LI>g.12345678_12345890dup</LI>\r\n</UL>", "VARCHAR(150)", "Genomic DNA change (HGVS format)|Description of variant at DNA level, based on the genomic DNA reference sequence (following HGVS recommendations); e.g. g.12345678C>T, g.12345678_12345890del, g.12345678_12345890dup.|text|30", "", "", 1, 1, 1, 0, NOW(), NULL, NULL)',
            'INSERT INTO ' . TABLE_COLS . ' VALUES ("VariantOnGenome/Frequency",                    9,  90, 0, 0, 0, "Frequency",            "", "Frequency in which the variant was found; e.g 5/760 chromosomes (in 5 of 760 chromosomes tested), 1/33 patients (in 1 of 33 patients analysed in study), 0.05 controls (in 5% of control cases tested).", "Frequency in which the variant was found; e.g 5/760 chromosomes (in 5 of 760 chromosomes tested), 1/33 patients (in 1 of 33 patients analysed in study), 0.05 controls (in 5% of control cases tested).", "VARCHAR(15)", "Frequency|Frequency in which the variant was found; e.g 5/760 chromosomes (in 5 of 760 chromosomes tested), 1/33 patients (in 1 of 33 patients analysed in study), 0.05 controls (in 5% of control cases tested). Preferred format is 3/75, not 0.04.|text|10", "", "", 1, 1, 1, 0, NOW(), NULL, NULL)',
            'INSERT INTO ' . TABLE_COLS . ' VALUES ("VariantOnGenome/Genetic_origin",              11, 200, 0, 0, 1, "Genetic origin",       "", "Origin of variant; unknown, germline (inherited), somatic, de novo, from parental disomy (maternal or paternal) or in vitro (cloned) when tested for functional consequences.", "Origin of variant; unknown, germline (inherited), somatic, de novo, from parental disomy (maternal or paternal) or in vitro (cloned) when tested for functional consequences.", "VARCHAR(100)", "Genetic origin||select|1|--Not specified--|false|false", "Unknown\r\nGermline (inherited)\r\nSomatic\r\nDe novo\r\nUniparental disomy\r\nUniparental disomy, maternal allele\r\nUniparental disomy, paternal allele\r\nIn vitro (cloned)", "", 1, 1, 1, 0, NOW(), NULL, NULL)',
            'INSERT INTO ' . TABLE_COLS . ' VALUES ("VariantOnGenome/Published_as",                 3, 200, 0, 0, 0, "Published as",         "Variant as originally reported (e.g. 521delT); provide only when different from \"DNA change\".", "Variant as originally reported (e.g. 521delT); listed only when different from \"DNA change\". Variants seen in animal models, tested in vitro, predicted from RNA analysis, etc. are described between brackets like c.(456C>G).", "Variant as originally reported (e.g. 521delT); listed only when different from \"DNA change\". Variants seen in animal models, tested in vitro, predicted from RNA analysis, etc. are described between brackets like c.(456C>G).", "VARCHAR(100)", "Published as|Variants seen in animal models, tested in vitro, predicted from RNA analysis, etc. are described between brackets like c.(456C>G).|text|30", "", "", 1, 1, 1, 0, NOW(), NULL, NULL)',
            'INSERT INTO ' . TABLE_COLS . ' VALUES ("VariantOnGenome/Reference",                    6, 200, 1, 1, 0, "Reference",            "", "Reference to publication describing the variant.", "Reference to publication describing the variant, including links to OMIM (when available), PubMed or or other source, e.g. \"den Dunnen ASHG2003 P2346\".", "VARCHAR(255)", "Reference||text|50", "", "", 1, 1, 1, 0, NOW(), NULL, NULL)',
            // Add remarks non public? Add remarks column(s) to VariantOnTranscript???
            'INSERT INTO ' . TABLE_COLS . ' VALUES ("VariantOnGenome/Remarks",                     10, 200, 0, 0, 1, "Variant remarks",      "Remarks regarding the variant described, e.g. germline mosaicism in mother, 345 kb deletion, muscle RNA analysed, not in 200 control chromosomes tested, on founder haplotype, etc.", "Remarks regarding the variant described.", "Remarks regarding the variant described, e.g. germline mosaicism in mother, 345 kb deletion, muscle RNA analysed, not in 200 control chromosomes tested, on founder haplotype, etc.", "TEXT", "Remarks||textarea|50|3", "", "", 1, 1, 1, 0, NOW(), NULL, NULL)',
            'INSERT INTO ' . TABLE_COLS . ' VALUES ("VariantOnGenome/Restriction_site",             5,  75, 0, 0, 0, "Re-site",              "", "Restriction enzyme recognition site created (+) or destroyed (-).", "Restriction enzyme recognition site created (+) or destroyed (-); e.g. BglII+, BamHI-.", "VARCHAR(25)", "Re-site|Restriction enzyme recognition site created (+) or destroyed (-); e.g. BglII+, BamHI-|text|10", "", "", 1, 1, 1, 0, NOW(), NULL, NULL)',
            'INSERT INTO ' . TABLE_COLS . ' VALUES ("VariantOnGenome/Segregation",                 12,  40, 0, 0, 0, "Segregation",          "Indicates whether the variant segregates with the phenotype (yes), does not segregate with the phenotype (no) or segregation is unknown (?)", "Indicates whether the variant segregates with the phenotype (yes), does not segregate with the phenotype (no) or segregation is unknown (?)", "Indicates whether the variant segregates with the phenotype (yes), does not segregate with the phenotype (no) or segregation is unknown (?)", "VARCHAR(100)", "Segregation||select|1|--Not specified--|false|false", "? = Unknown\r\nyes = Segregates with phenotype\r\nno = Does not segregate with phenotype", "", 1, 1, 1, 0, NOW(), NULL, NULL)',
            // FIXME; link this one to an ontology?
            'INSERT INTO ' . TABLE_COLS . ' VALUES ("VariantOnGenome/Type",                         1, 200, 0, 0, 1, "Type",                 "", "Type of variant at DNA level.", "Type of variant at DNA level; note that the variant type can also be derived from the variant description (for all levels).", "VARCHAR(100)", "Type of variant (DNA level)|Type of variant at DNA level; note that the variant type can also be derived from the variant description (for all levels).|select|1|true|false|false", "Substitution\r\nDeletion\r\nDuplication\r\nInsertion\r\nInversion\r\nInsertion/Deletion\r\nTranslocation\r\nOther/Complex", "", 1, 1, 1, 0, NOW(), NULL, NULL)',
            'INSERT INTO ' . TABLE_COLS . ' VALUES ("VariantOnTranscript/Distance_to_splice_site", 10, 150, 0, 1, 0, "Splice distance",      "", "The distance to the nearest splice site.", "The distance to the nearest splice site.", "MEDIUMINT(8) UNSIGNED", "Distance to splice site||text|8", "", "", 0, 0, 1, 0, NOW(), NULL, NULL)',
            'INSERT INTO ' . TABLE_COLS . ' VALUES ("VariantOnTranscript/DNA",                      3, 200, 1, 1, 1, "DNA change (cDNA)",    "", "Description of variant at DNA level, based on a coding DNA reference sequence (following HGVS recommendations).", "Description of variant at DNA level, based on a coding DNA reference sequence (following HGVS recommendations); e.g. c.123C>T, c.123_145del, c.123_126dup.", "VARCHAR(100)", "DNA change (HGVS format)|Description of variant at DNA level, based on a coding DNA reference sequence (following HGVS recommendations); e.g. c.123C>T, c.123_145del, c.123_126dup.|text|30", "", "", 1, 1, 1, 0, NOW(), NULL, NULL)',
            'INSERT INTO ' . TABLE_COLS . ' VALUES ("VariantOnTranscript/Exon",                     2,  50, 0, 0, 1, "Exon",                 "", "Number of exon/intron containing the variant.", "Number of exon/intron containing variant; 2 = exon 2, 12i = intron 12, 2i_7i = exons 3 to 7, 8i_9 = border intron 8/exon 9.", "VARCHAR(7)", "Exon|Format: 2 = exon 2, 12i = intron 12, 2i_7i = exons 3 to 7, 8i_9 = border intron 8/exon 9.|text|7", "", "", 1, 1, 1, 0, NOW(), NULL, NULL)',
            'INSERT INTO ' . TABLE_COLS . ' VALUES ("VariantOnTranscript/GVS/Function",             9, 200, 0, 1, 0, "GVS function",         "", "Functional annotation of this position from the Genome Variation Server.", "The functional annotation of this position from the Genome Variation Server.", "VARCHAR(100)", "GVS function||select|1|true|false|false", "intergenic\r\nnear-gene-5\r\nutr-5\r\ncoding\r\ncoding-near-splice\r\ncoding-synonymous\r\ncoding-synonymous-near-splice\r\ncodingComplex\r\ncodingComplex-near-splice\r\nframeshift\r\nframeshift-near-splice\r\nmissense\r\nmissense-near-splice\r\nsplice-5\r\nsplice\r\nintron\r\nsplice-3\r\nstop-gained\r\nstop-gained-near-splice\r\nstop-lost\r\nstop-lost-near-splice\r\nutr-3\r\nnear-gene-3", "", 1, 1, 1, 0, NOW(), NULL, NULL)',
            'INSERT INTO ' . TABLE_COLS . ' VALUES ("VariantOnTranscript/Location",                 1, 200, 0, 0, 1, "Location",             "", "Location of variant at DNA level.", "Location of variant at DNA level; note that the variant location can also be derived from the variant description.", "VARCHAR(100)", "Location of variant|The variant location can also be derived from the variant description|select|1|true|false|false", "5\' gene flanking\r\n5\' UTR\r\nExon\r\nIntron\r\n3\' UTR\r\n3\' gene flanking", "", 1, 1, 1, 0, NOW(), NULL, NULL)',
            'INSERT INTO ' . TABLE_COLS . ' VALUES ("VariantOnTranscript/PolyPhen",                 8, 200, 0, 0, 0, "PolyPhen prediction",  "", "Effect of variant, predicted by PolyPhen.", "Effect of variant, predicted by PolyPhen.", "VARCHAR(100)", "PolyPhen prediction|Effect of variant, predicted by PolyPhen|select|1|true|false|false", "benign = Benign\r\npossiblyDamaging = Possibly damaging\r\nprobablyDamaging = Probably damaging\r\nnoPrediction = No prediction", "", 1, 1, 1, 0, NOW(), NULL, NULL)',
            'INSERT INTO ' . TABLE_COLS . ' VALUES ("VariantOnTranscript/Position",                 5, 100, 0, 1, 0, "Position",             "", "Position of variant in coding DNA sequence.", "Position of variant in coding DNA sequence; note that coding DNA position can also be derived from the variant description.", "MEDIUMINT(6)", "cDNA position|Note that cDNA position can be derived from the variant description.|text|5", "", "", 1, 1, 1, 0, NOW(), NULL, NULL)',
            'INSERT INTO ' . TABLE_COLS . ' VALUES ("VariantOnTranscript/Protein",                  7, 200, 1, 1, 1, "Protein",              "", "Description of variant at protein level (following HGVS recommendations).", "Description of variant at protein level (following HGVS recommendations).<BR>\r\n<UL style=\"margin-top : 0px;\">\r\n  <LI>p.(Arg345Pro) = change predicted from DNA (RNA not analysed)</LI>\r\n  <LI>p.Arg345Pro = change derived from RNA analysis</LI>\r\n  <LI>p.? = unknown effect</LI>\r\n  <LI>p.0? = probably no protein produced</LI>\r\n</UL>", "VARCHAR(100)", "Protein change (HGVS format)|Description of variant at protein level (following HGVS recommendations); e.g. p.(Arg345Pro) = change predicted from DNA (RNA not analysed), p.Arg345Pro = change derived from RNA analysis, p.0 (no protein produced), p.? (unknown effect).|text|30", "", "", 1, 1, 1, 0, NOW(), NULL, NULL)',
            'INSERT INTO ' . TABLE_COLS . ' VALUES ("VariantOnTranscript/Published_as",             4, 200, 0, 0, 0, "Published as",         "Variant as originally reported (e.g. 521delT); provide only when different from \"DNA change\".", "Variant as originally reported (e.g. 521delT); listed only when different from \"DNA change\". Variants seen in animal models, tested in vitro, predicted from RNA analysis, etc. are described between brackets like c.(456C>G).", "Variant as originally reported (e.g. 521delT); listed only when different from \"DNA change\". Variants seen in animal models, tested in vitro, predicted from RNA analysis, etc. are described between brackets like c.(456C>G).", "VARCHAR(100)", "Published as|Variants seen in animal models, tested in vitro, predicted from RNA analysis, etc. are described between brackets like c.(456C>G).|text|30", "", "", 1, 1, 1, 0, NOW(), NULL, NULL)',
            'INSERT INTO ' . TABLE_COLS . ' VALUES ("VariantOnTranscript/RNA",                      6, 200, 1, 1, 1, "RNA change",           "", "Description of variant at RNA level (following HGVS recommendations).", "Description of variant at RNA level (following HGVS recommendations).<BR>\r\n<UL style=\"margin-top : 0px;\">\r\n  <LI>r.123c>u</LI>\r\n  <LI>r.? = unknown</LI>\r\n  <LI>r.(?) = RNA not analysed but probably transcribed copy of DNA variant</LI>\r\n  <LI>r.spl? = RNA not analysed but variant probably affects splicing</LI>\r\n  <LI>r.(spl?) = RNA not analysed but variant may affect splicing</LI>\r\n  <LI>r.0? = change expected to abolish transcription</LI>\r\n</UL>", "VARCHAR(100)", "RNA change (HGVS format)|Description of variant at RNA level (following HGVS recommendations); e.g. r.123c>u, r.? = unknown, r.(?) = RNA not analysed but probably transcribed copy of DNA variant, r.spl? = RNA not analysed but variant probably affects splicing, r.(spl?) = RNA not analysed but variant may affect splicing.|text|30", "", "", 1, 1, 1, 0, NOW(), NULL, NULL)',





            'INSERT INTO ' . TABLE_COLS . ' VALUES ("Screening/Analysis_restricted",                        255, 100, 0, 1, 0, "Analysis restricted", "", "Should the analysis be restricted to the gene panel?", "Should the analysis be restricted to the gene panel?", "BOOLEAN", "Analysis restricted||checkbox", "", "", 1, 0, 1, 0, NOW(), NULL, NULL)',
            'INSERT INTO ' . TABLE_COLS . ' VALUES ("Screening/Covered_exome/Fraction",                     255, 100, 0, 1, 0, "Exome covered", "", "The amount of exome covered.", "The amount of exome covered by this sequencing run.", "FLOAT UNSIGNED", "Exome covered||text|6", "", "", 1, 0, 1, 0, NOW(), NULL, NULL)',
            'INSERT INTO ' . TABLE_COLS . ' VALUES ("Screening/Derived_gender",                             255, 100, 0, 1, 0, "Derived gender", "", "Derived gender of individual.", "Derived gender of individual.", "VARCHAR(100)", "Derived gender||select|1|--Not specified--|false|false", "? = Unknown\r\nF = Female\r\nM = Male\r\nrF = Raised as female\r\nrM = Raised as male", "", 1, 0, 1, 0, NOW(), NULL, NULL)',
            'INSERT INTO ' . TABLE_COLS . ' VALUES ("Screening/Father/Covered_exome/Fraction",              255, 100, 0, 1, 0, "Exome covered (father)", "", "The amount of exome covered (paternal).", "The amount of exome covered by this sequencing run (paternal).", "FLOAT UNSIGNED", "Exome covered (father)||text|6", "", "", 1, 0, 1, 0, NOW(), NULL, NULL)',
            'INSERT INTO ' . TABLE_COLS . ' VALUES ("Screening/Father/Panel_coverage/Fraction",             255, 100, 0, 1, 0, "Panel coverage (father)", "", "The amount of coverage in the panel (paternal).", "The amount of coverage in the panel in this sequencing run (paternal).", "FLOAT UNSIGNED", "Panel coverage (father)||text|6", "", "", 1, 0, 1, 0, NOW(), NULL, NULL)',
            'INSERT INTO ' . TABLE_COLS . ' VALUES ("Screening/Father/Reads_on_target/Fraction",            255, 100, 0, 1, 0, "Reads on target (father)", "", "The amount of reads on target (paternal).", "The amount of reads on target in this sequencing run (paternal).", "FLOAT UNSIGNED", "Reads on target (father)||text|6", "", "", 1, 0, 1, 0, NOW(), NULL, NULL)',
            'INSERT INTO ' . TABLE_COLS . ' VALUES ("Screening/Mother/Covered_exome/Fraction",              255, 100, 0, 1, 0, "Exome covered (mother)", "", "The amount of exome covered (maternal).", "The amount of exome covered by this sequencing run (maternal).", "FLOAT UNSIGNED", "Exome covered (mother)||text|6", "", "", 1, 0, 1, 0, NOW(), NULL, NULL)',
            'INSERT INTO ' . TABLE_COLS . ' VALUES ("Screening/Mother/Panel_coverage/Fraction",             255, 100, 0, 1, 0, "Panel coverage (mother)", "", "The amount of coverage in the panel (maternal).", "The amount of coverage in the panel in this sequencing run (maternal).", "FLOAT UNSIGNED", "Panel coverage (mother)||text|6", "", "", 1, 0, 1, 0, NOW(), NULL, NULL)',
            'INSERT INTO ' . TABLE_COLS . ' VALUES ("Screening/Mother/Reads_on_target/Fraction",            255, 100, 0, 1, 0, "Reads on target (mother)", "", "The amount of reads on target (maternal).", "The amount of reads on target in this sequencing run (maternal).", "FLOAT UNSIGNED", "Reads on target (mother)||text|6", "", "", 1, 0, 1, 0, NOW(), NULL, NULL)',
            'INSERT INTO ' . TABLE_COLS . ' VALUES ("Screening/Panel_coverage/Fraction",                    255, 100, 0, 1, 0, "Panel coverage", "", "The amount of coverage in the panel.", "The amount of coverage in the panel in this sequencing run.", "FLOAT UNSIGNED", "Panel coverage||text|6", "", "", 1, 0, 1, 0, NOW(), NULL, NULL)',
            'INSERT INTO ' . TABLE_COLS . ' VALUES ("Screening/Reads_on_target/Fraction",                   255, 100, 0, 1, 0, "Reads on target", "", "The amount of reads on target.", "The amount of reads on target in this sequencing run.", "FLOAT UNSIGNED", "Reads on target||text|6", "", "", 1, 0, 1, 0, NOW(), NULL, NULL)',
            'INSERT INTO ' . TABLE_COLS . ' VALUES ("Screening/SNP_overlap",                                255, 100, 0, 1, 0, "SNP overlap", "", "Overlap of found variants with the SNP array method.", "Overlap of found variants with the SNP array method.", "FLOAT UNSIGNED", "SNP overlap||text|6", "", "", 1, 0, 1, 0, NOW(), NULL, NULL)',
            'INSERT INTO ' . TABLE_COLS . ' VALUES ("Screening/Trio_check/De_novo",                         255, 100, 0, 1, 0, "Trio check: De novo", "", "The number of de novo variants reported by the trio check.", "The number of de novo variants reported by the trio check.", "SMALLINT UNSIGNED", "Trio check: De novo||text|6", "", "", 1, 0, 1, 0, NOW(), NULL, NULL)',
            'INSERT INTO ' . TABLE_COLS . ' VALUES ("Screening/Trio_check/Mendelian",                       255, 100, 0, 1, 0, "Trio check: Mendelian", "", "The percentage of mendelian inheritance, as reported by the trio check.", "The percentage of mendelian inheritance, as reported by the trio check.", "FLOAT UNSIGNED", "Trio check: Mendelian||text|6", "", "", 1, 0, 1, 0, NOW(), NULL, NULL)',
            'INSERT INTO ' . TABLE_COLS . ' VALUES ("VariantOnGenome/Alamut",                               255, 100, 0, 1, 0, "Alamut link", "", "Links to the variant in Alamut.", "Links to the variant in Alamut.", "VARCHAR(255)", "Alamut link||text|20", "", "", 1, 1, 1, 0, NOW(), NULL, NULL)',
            'INSERT INTO ' . TABLE_COLS . ' VALUES ("VariantOnGenome/CADD/Raw",                             255, 100, 0, 1, 0, "CADD Raw", "", "CADD Raw variant effect score.", "CADD Raw variant effect score.", "FLOAT", "CADD Raw||text|6", "", "", 0, 0, 1, 0, NOW(), NULL, NULL)',
            'INSERT INTO ' . TABLE_COLS . ' VALUES ("VariantOnGenome/CADD/Phred",                           255, 100, 0, 1, 0, "CADD Phred", "", "CADD Phred variant effect score.", "CADD Phred variant effect score.", "FLOAT UNSIGNED", "CADD Phred||text|6", "", "", 0, 0, 1, 0, NOW(), NULL, NULL)',
            'INSERT INTO ' . TABLE_COLS . ' VALUES ("VariantOnGenome/Conservation_score/Phast",             255, 100, 0, 1, 0, "Phast conservation", "", "Conservation score as calculated by Phast.", "Conservation score as calculated by Phast.", "FLOAT UNSIGNED", "Phast conservation score||text|6", "", "", 0, 0, 1, 0, NOW(), NULL, NULL)',
            'INSERT INTO ' . TABLE_COLS . ' VALUES ("VariantOnGenome/Conservation_score/PhyloP",            255, 100, 0, 1, 0, "PhyloP conservation", "", "Conservation score as calculated by PhyloP.", "The Conservation score as calculated by PhyloP.", "FLOAT", "PhyloP conservation score||text|6", "", "", 1, 1, 1, 0, NOW(), NULL, NULL)',
            'INSERT INTO ' . TABLE_COLS . ' VALUES ("VariantOnGenome/Frequency/1000G",                      255, 100, 0, 1, 0, "1000G AF", "", "Allele frequency from the 1000 Genomes project.", "Allele frequency from the 1000 Genomes project.", "FLOAT UNSIGNED", "1000G allele frequency||text|6", "", "", 0, 0, 1, 0, NOW(), NULL, NULL)',
            'INSERT INTO ' . TABLE_COLS . ' VALUES ("VariantOnGenome/Frequency/GoNL",                       255, 100, 0, 1, 0, "GoNL AF", "", "Allele frequency from the GoNL project.", "Allele frequency from the GoNL project.", "FLOAT UNSIGNED", "GoNL allele frequency||text|6", "", "", 0, 0, 1, 0, NOW(), NULL, NULL)',
            'INSERT INTO ' . TABLE_COLS . ' VALUES ("VariantOnGenome/Frequency/EVS",                        255, 100, 0, 1, 0, "EVS AF", "", "Allele frequency from the EVS project.", "Allele frequency from the EVS project.", "FLOAT UNSIGNED", "EVS allele frequency||text|6", "", "", 0, 0, 1, 0, NOW(), NULL, NULL)',
            'INSERT INTO ' . TABLE_COLS . ' VALUES ("VariantOnGenome/HGMD/Association",                     255, 100, 0, 1, 0, "HGMD association", "", "Association with disease according to HGMD.", "Association with disease according to HGMD.", "VARCHAR(255)", "HGMD association||text|20", "", "", 0, 0, 1, 0, NOW(), NULL, NULL)',
            'INSERT INTO ' . TABLE_COLS . ' VALUES ("VariantOnGenome/HGMD/Reference",                       255, 100, 0, 1, 0, "HGMD reference", "", "Reference for association with disease according to HGMD.", "Reference for association with disease according to HGMD.", "VARCHAR(255)", "HGMD reference||text|20", "", "", 0, 0, 1, 0, NOW(), NULL, NULL)',
            'INSERT INTO ' . TABLE_COLS . ' VALUES ("VariantOnGenome/InhouseDB/Count/HC",                   255, 100, 0, 1, 0, "INDB Count HC", "", "The count of this variant in the Inhouse Database (HC).", "The count of this variant in the Inhouse Database (HC).", "SMALLINT UNSIGNED", "INDB Count HC||text|6", "", "", 0, 0, 1, 0, NOW(), NULL, NULL)',
            'INSERT INTO ' . TABLE_COLS . ' VALUES ("VariantOnGenome/InhouseDB/Count/UG",                   255, 100, 0, 1, 0, "INDB Count UG", "", "The count of this variant in the Inhouse Database (UG).", "The count of this variant in the Inhouse Database (UG).", "SMALLINT UNSIGNED", "INDB Count UG||text|6", "", "", 0, 0, 1, 0, NOW(), NULL, NULL)',
            'INSERT INTO ' . TABLE_COLS . ' VALUES ("VariantOnGenome/Sequencing/Quality",                   255, 100, 0, 1, 0, "Sequencing quality", "", "Sequencing quality.", "Sequencing quality, as presented in the input file.", "FLOAT UNSIGNED", "Sequencing quality||text|6", "", "", 0, 0, 1, 0, NOW(), NULL, NULL)',
            'INSERT INTO ' . TABLE_COLS . ' VALUES ("VariantOnGenome/Sequencing/Filter",                    255, 100, 0, 1, 0, "Sequencing filter", "", "Sequencing filter.", "Sequencing filter output, as presented in the input file.", "VARCHAR(50)", "Sequencing filter||text|40", "", "", 0, 0, 1, 0, NOW(), NULL, NULL)',
            'INSERT INTO ' . TABLE_COLS . ' VALUES ("VariantOnGenome/Sequencing/GATKcaller",                255,  80, 0, 1, 0, "GATKcaller", "", "GATKcaller.", "GATKcaller.", "VARCHAR(8)", "GATKcaller||text|10", "", "", 0, 0, 1, 0, NOW(), NULL, NULL)',
            'INSERT INTO ' . TABLE_COLS . ' VALUES ("VariantOnGenome/Sequencing/Depth/Total",               255, 100, 0, 1, 0, "Read depth", "", "Sequencing read depth.", "Sequencing read depth, as presented in the input file.", "SMALLINT UNSIGNED", "Read depth||text|6", "", "", 0, 0, 1, 0, NOW(), NULL, NULL)',
            'INSERT INTO ' . TABLE_COLS . ' VALUES ("VariantOnGenome/Sequencing/Depth/Alt",                 255, 100, 0, 1, 0, "Read depth Alt", "", "Sequencing read depth of Alt allele.", "Sequencing read depth of Alt allele, as presented in the input file.", "SMALLINT UNSIGNED", "Read depth Alt||text|6", "", "", 0, 0, 1, 0, NOW(), NULL, NULL)',
            'INSERT INTO ' . TABLE_COLS . ' VALUES ("VariantOnGenome/Sequencing/Depth/Alt/Fraction",        255, 100, 0, 1, 0, "Read depth Alt (fraction)", "", "Sequencing read depth of Alt allele as a fraction.", "Sequencing read depth of Alt allele, as presented in the input file, as the fraction of the total number of reads of Ref and Alt allele together.", "FLOAT UNSIGNED", "Read depth Alt (fraction)||text|6", "", "", 0, 0, 1, 0, NOW(), NULL, NULL)',
            'INSERT INTO ' . TABLE_COLS . ' VALUES ("VariantOnGenome/Sequencing/Depth/Ref",                 255, 100, 0, 1, 0, "Read depth Ref", "", "Sequencing read depth of Ref allele.", "Sequencing read depth of Ref allele, as presented in the input file.", "SMALLINT UNSIGNED", "Read depth Ref||text|6", "", "", 0, 0, 1, 0, NOW(), NULL, NULL)',
            'INSERT INTO ' . TABLE_COLS . ' VALUES ("VariantOnGenome/Sequencing/Father/Depth/Total",        255, 100, 0, 1, 0, "Read depth father", "", "Sequencing read depth of unaffected father.", "Sequencing read depth of unaffected father, as presented in the input file.", "SMALLINT UNSIGNED", "Read depth father||text|6", "", "", 0, 0, 1, 0, NOW(), NULL, NULL)',
            'INSERT INTO ' . TABLE_COLS . ' VALUES ("VariantOnGenome/Sequencing/Mother/Depth/Total",        255, 100, 0, 1, 0, "Read depth mother", "", "Sequencing read depth of unaffected mother.", "Sequencing read depth of unaffected mother, as presented in the input file.", "SMALLINT UNSIGNED", "Read depth mother||text|6", "", "", 0, 0, 1, 0, NOW(), NULL, NULL)',
            'INSERT INTO ' . TABLE_COLS . ' VALUES ("VariantOnGenome/Sequencing/Father/GenoType",           255, 100, 0, 1, 0, "Genotype of father", "", "The genotype of the unaffected father.", "The genotype of the unaffected father.", "VARCHAR(3)", "Genotype of father||text|4", "", "", 0, 0, 1, 0, NOW(), NULL, NULL)',
            'INSERT INTO ' . TABLE_COLS . ' VALUES ("VariantOnGenome/Sequencing/Father/VarPresent",         255, 100, 0, 1, 0, "Variant present in father", "", "Is the variant present in the unaffected father? Scores 1-6.", "Is the variant present in the unaffected father? Scores 1-6. 1 certainly not present, 6 certainly present.", "TINYINT UNSIGNED", "Variant present in father||select|1|true|false|false", "1=Not present with DP>20\r\n2=Not present with DP 8-20\r\n3=Not present with DP<8\r\n4=Present with <10% of reads\r\n5=Present with >10% of reads\r\n6=Called by pipeline", "", 0, 0, 1, 0, NOW(), NULL, NULL)',
            'INSERT INTO ' . TABLE_COLS . ' VALUES ("VariantOnGenome/Sequencing/Father/Depth/Alt/Fraction", 255, 100, 0, 1, 0, "Read depth father Alt (fraction)", "", "Sequencing read depth of Alt allele of unaffected father as a fraction.", "Sequencing read depth of Alt allele of unaffected father, as presented in the input file, as the fraction of the total number of reads of Ref and Alt allele together.", "FLOAT UNSIGNED", "Read depth father Alt (fraction)||text|6", "", "", 0, 0, 1, 0, NOW(), NULL, NULL)',
            'INSERT INTO ' . TABLE_COLS . ' VALUES ("VariantOnGenome/Sequencing/Mother/GenoType",           255, 100, 0, 1, 0, "Genotype of mother", "", "The genotype of the unaffected mother.", "The genotype of the unaffected mother.", "VARCHAR(3)", "Genotype of mother||text|4", "", "", 0, 0, 1, 0, NOW(), NULL, NULL)',
            'INSERT INTO ' . TABLE_COLS . ' VALUES ("VariantOnGenome/Sequencing/Mother/VarPresent",         255, 100, 0, 1, 0, "Variant present in mother", "", "Is the variant present in the unaffected mother? Scores 1-6.", "Is the variant present in the unaffected mother? Scores 1-6. 1 certainly not present, 6 certainly present.", "TINYINT UNSIGNED", "Variant present in mother||select|1|true|false|false", "1=Not present with DP>20\r\n2=Not present with DP 8-20\r\n3=Not present with DP<8\r\n4=Present with <10% of reads\r\n5=Present with >10% of reads\r\n6=Called by pipeline", "", 0, 0, 1, 0, NOW(), NULL, NULL)',
            'INSERT INTO ' . TABLE_COLS . ' VALUES ("VariantOnGenome/Sequencing/Mother/Depth/Alt/Fraction", 255, 100, 0, 1, 0, "Read depth mother Alt (fraction)", "", "Sequencing read depth of Alt allele of unaffected mother as a fraction.", "Sequencing read depth of Alt allele of unaffected mother, as presented in the input file, as the fraction of the total number of reads of Ref and Alt allele together.", "FLOAT UNSIGNED", "Read depth mother Alt (fraction)||text|6", "", "", 0, 0, 1, 0, NOW(), NULL, NULL)',
            'INSERT INTO ' . TABLE_COLS . ' VALUES ("VariantOnTranscript/PolyPhen/HDIV",                      8, 200, 0, 1, 0, "PolyPhen HDIV",  "", "Effect of variant, predicted by PolyPhen, HDIV score.", "Effect of variant, predicted by PolyPhen, HDIV score.", "VARCHAR(200)", "PolyPhen prediction, HDIV score||text|5", "", "", 0, 0, 1, 0, NOW(), NULL, NULL)',
            'INSERT INTO ' . TABLE_COLS . ' VALUES ("VariantOnTranscript/PolyPhen/HVAR",                      8, 200, 0, 1, 0, "PolyPhen HVAR",  "", "Effect of variant, predicted by PolyPhen, HVAR score.", "Effect of variant, predicted by PolyPhen, HVAR score.", "VARCHAR(250)", "PolyPhen prediction, HVAR score||text|5", "", "", 0, 0, 1, 0, NOW(), NULL, NULL)',
            'INSERT INTO ' . TABLE_COLS . ' VALUES ("VariantOnTranscript/Prediction/MutationTaster",        255, 100, 0, 1, 0, "MutationTaster prediction", "", "MutationTaster prediction.", "The MutationTaster prediction, predicting the effect of the DNA variant on the function of the protein.", "CHAR(1)", "MutationTaster prediction||select|1|true|false|false", "A\r\nD\r\nN\r\nP", "", 0, 0, 1, 0, NOW(), NULL, NULL)',
            'INSERT INTO ' . TABLE_COLS . ' VALUES ("VariantOnTranscript/Prediction/MutationTaster/Score",  255, 100, 0, 1, 0, "MutationTaster score", "", "MutationTaster score.", "The MutationTaster score, predicting the effect of the DNA variant on the function of the protein.", "FLOAT", "MutationTaster score||text|6", "", "", 0, 0, 1, 0, NOW(), NULL, NULL)',
            'INSERT INTO ' . TABLE_COLS . ' VALUES ("VariantOnTranscript/Prediction/SIFT",                  255, 100, 0, 1, 0, "SIFT score", "", "SIFT score.", "The SIFT score, predicting the effect of the DNA variant on the function of the protein.", "VARCHAR(50)", "SIFT score||text|6", "", "", 0, 0, 1, 0, NOW(), NULL, NULL)',
            'INSERT INTO ' . TABLE_COLS . ' VALUES ("VariantOnTranscript/Prediction/Grantham",              255, 100, 0, 1, 0, "Grantham score", "", "Grantham score.", "The Grantham score, predicting the effect of the DNA variant on the function of the protein.", "TINYINT UNSIGNED", "Grantham score||text|6", "", "", 0, 0, 1, 0, NOW(), NULL, NULL)',
        );



}

// DMD_SPECIFIC;
if (lovd_getProjectFile() == '/install/inc-sql-columns.php') {
    header('Content-type: text/plain; charset=UTF-8');
    print_r($aMGHAColSQL);
}
?>
